{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "23618c5d35fa4d42938a8037c995fd14": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7c25053baff7404bbcab92364d61cd05",
              "IPY_MODEL_aa62410d6737451187ff528af39a37e9",
              "IPY_MODEL_f7cb75813e874fa69f81be93c3fac28d"
            ],
            "layout": "IPY_MODEL_12f97800aacd4de7af630f64b535db7b"
          }
        },
        "7c25053baff7404bbcab92364d61cd05": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_856434b2de6d4f749895c07156094209",
            "placeholder": "​",
            "style": "IPY_MODEL_bfffee1ab7d54907bb4b962a2d652afb",
            "value": ""
          }
        },
        "aa62410d6737451187ff528af39a37e9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_febd2c6447af46f1bb1aeb0c53efafc0",
            "max": 1,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_210407cf4be94e99a15b650723167f0d",
            "value": 0
          }
        },
        "f7cb75813e874fa69f81be93c3fac28d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_74851daed77f493d81cadd73f3969b42",
            "placeholder": "​",
            "style": "IPY_MODEL_9029b80b30c74a6c9bed95baa5545b20",
            "value": " 0/0 [00:00&lt;?, ?it/s]"
          }
        },
        "12f97800aacd4de7af630f64b535db7b": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "856434b2de6d4f749895c07156094209": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "bfffee1ab7d54907bb4b962a2d652afb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "febd2c6447af46f1bb1aeb0c53efafc0": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": "20px"
          }
        },
        "210407cf4be94e99a15b650723167f0d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "74851daed77f493d81cadd73f3969b42": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9029b80b30c74a6c9bed95baa5545b20": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "80f0b7d8bfcb40f78ce491ed227761e8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_904776d93faa44958ba68487e87251e7",
              "IPY_MODEL_723b413007b242918506390013ee0541",
              "IPY_MODEL_7f0ee27686fa4c88b866d37893777978"
            ],
            "layout": "IPY_MODEL_0de26fcbd73e4eadab35eb21c3faefb3"
          }
        },
        "904776d93faa44958ba68487e87251e7": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_0cb06b65aff14f38907b4f06e5d362be",
            "placeholder": "​",
            "style": "IPY_MODEL_4f5bc8679cc7447ea3932e37fd92e7a2",
            "value": "Downloading: 100%"
          }
        },
        "723b413007b242918506390013ee0541": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_0c82f3ff048e40c1838ec1f25e382d12",
            "max": 1802,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_b74502b83e5149b385c60dfd6ec00d4a",
            "value": 1802
          }
        },
        "7f0ee27686fa4c88b866d37893777978": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_65fb1c906fc04687b460a5fe5b78a666",
            "placeholder": "​",
            "style": "IPY_MODEL_b838d14e85fa4ebdbfd88a662c0a7540",
            "value": " 1.80k/1.80k [00:00&lt;00:00, 52.5kB/s]"
          }
        },
        "0de26fcbd73e4eadab35eb21c3faefb3": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "0cb06b65aff14f38907b4f06e5d362be": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "4f5bc8679cc7447ea3932e37fd92e7a2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "0c82f3ff048e40c1838ec1f25e382d12": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b74502b83e5149b385c60dfd6ec00d4a": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "65fb1c906fc04687b460a5fe5b78a666": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b838d14e85fa4ebdbfd88a662c0a7540": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "f84f090c0dec465c8a9490e1f50e7e74": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_1c20e9f6f07046d89dac6a186402fb7c",
              "IPY_MODEL_f4c9c81440f3467587dfab9dac2cf9eb",
              "IPY_MODEL_26d4d2b134ec4a5baf3571668fc2bf84"
            ],
            "layout": "IPY_MODEL_c8ad5be4d12b4cdd881fe0e17d5f78e1"
          }
        },
        "1c20e9f6f07046d89dac6a186402fb7c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_174725c73a9f4f2e87b758772a5936af",
            "placeholder": "​",
            "style": "IPY_MODEL_a54813923ec04ed0babc76f192d40027",
            "value": "Downloading: 100%"
          }
        },
        "f4c9c81440f3467587dfab9dac2cf9eb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_fa48ac996102414b909d6fed0a28912f",
            "max": 1222317369,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_eb2194b21480456eb72b2edc3d35568e",
            "value": 1222317369
          }
        },
        "26d4d2b134ec4a5baf3571668fc2bf84": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_5733a9722b24405ba258cba1b7de50a1",
            "placeholder": "​",
            "style": "IPY_MODEL_0b76ad47a8c84771b855b4d9fa780adb",
            "value": " 1.22G/1.22G [00:25&lt;00:00, 51.3MB/s]"
          }
        },
        "c8ad5be4d12b4cdd881fe0e17d5f78e1": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "174725c73a9f4f2e87b758772a5936af": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "a54813923ec04ed0babc76f192d40027": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "fa48ac996102414b909d6fed0a28912f": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "eb2194b21480456eb72b2edc3d35568e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "5733a9722b24405ba258cba1b7de50a1": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "0b76ad47a8c84771b855b4d9fa780adb": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "38fdfbc7ae7b4fc096345eb2bbb4fef9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7c162c141fdf4330bec85b4e2c50bbbf",
              "IPY_MODEL_0faf0b13b1174d5086953cb85d7bcb23",
              "IPY_MODEL_eb506ea11331493886ae62a45342960b"
            ],
            "layout": "IPY_MODEL_ad2baddab175472c8d0df25bbddafdd7"
          }
        },
        "7c162c141fdf4330bec85b4e2c50bbbf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_25ab995d3fea419f8bcd024363d87fbb",
            "placeholder": "​",
            "style": "IPY_MODEL_d7375a7afe6f406a98c414641a9f8abe",
            "value": "Downloading: 100%"
          }
        },
        "0faf0b13b1174d5086953cb85d7bcb23": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_bef5a9799baa4127b5afa9c2c0b97f70",
            "max": 26,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_f99af7c7746d4a0bb4d9576a27943c57",
            "value": 26
          }
        },
        "eb506ea11331493886ae62a45342960b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_35684ce6f4d54c33800041aef11e9203",
            "placeholder": "​",
            "style": "IPY_MODEL_dea024743e2c4f1b9509481411522940",
            "value": " 26.0/26.0 [00:00&lt;00:00, 757B/s]"
          }
        },
        "ad2baddab175472c8d0df25bbddafdd7": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "25ab995d3fea419f8bcd024363d87fbb": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d7375a7afe6f406a98c414641a9f8abe": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "bef5a9799baa4127b5afa9c2c0b97f70": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "f99af7c7746d4a0bb4d9576a27943c57": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "35684ce6f4d54c33800041aef11e9203": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "dea024743e2c4f1b9509481411522940": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "e58939836cd24adaade2f6b77c2260f8": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7268b6912e8a4de39c0deb7955649262",
              "IPY_MODEL_5e39735d65c8460aad3ade749330320e",
              "IPY_MODEL_97df69b0220646b7957d9c0fbd94d2f2"
            ],
            "layout": "IPY_MODEL_030c740e71894835a2a1e6dd60caa853"
          }
        },
        "7268b6912e8a4de39c0deb7955649262": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d84a371009b249cb922ab5d4e8688718",
            "placeholder": "​",
            "style": "IPY_MODEL_89eeb246e8c94af7a665d2a1a2fbd326",
            "value": "Downloading: 100%"
          }
        },
        "5e39735d65c8460aad3ade749330320e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_67e201ede8414ba98731a481ed4c6128",
            "max": 898822,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_6257c8fc4411421fba4d38f5dc015b42",
            "value": 898822
          }
        },
        "97df69b0220646b7957d9c0fbd94d2f2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_9d15cb0f3a1e4e09970408bccda8a979",
            "placeholder": "​",
            "style": "IPY_MODEL_7aed8bb753bb487e8fa951b147804921",
            "value": " 899k/899k [00:00&lt;00:00, 11.4MB/s]"
          }
        },
        "030c740e71894835a2a1e6dd60caa853": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d84a371009b249cb922ab5d4e8688718": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "89eeb246e8c94af7a665d2a1a2fbd326": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "67e201ede8414ba98731a481ed4c6128": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "6257c8fc4411421fba4d38f5dc015b42": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "9d15cb0f3a1e4e09970408bccda8a979": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7aed8bb753bb487e8fa951b147804921": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "a5e856ba81ce47edaf5fcce079b75aa5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_398a521c850341e390f35299c68ca316",
              "IPY_MODEL_05987e78a5764fb99e31fed747ba1a86",
              "IPY_MODEL_dd72772cd6ca4f5a8c3167fc933cafca"
            ],
            "layout": "IPY_MODEL_d970095f0b9a477aa20b54b0c28b4f41"
          }
        },
        "398a521c850341e390f35299c68ca316": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_3db97d36de3a4f1e989350a37c6a286e",
            "placeholder": "​",
            "style": "IPY_MODEL_54954ee923594839b674c422e552fcce",
            "value": "Downloading: 100%"
          }
        },
        "05987e78a5764fb99e31fed747ba1a86": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_6e9fb008b2ff4ed0892882dc8aa78f88",
            "max": 456318,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_a7ac8db34f9340ea994ed66f8823a300",
            "value": 456318
          }
        },
        "dd72772cd6ca4f5a8c3167fc933cafca": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_624ecab9cbe64c3e8c8b2ed96733d3fa",
            "placeholder": "​",
            "style": "IPY_MODEL_d8f5c0d33a8c4d5f91164a54236e3e47",
            "value": " 456k/456k [00:00&lt;00:00, 7.69MB/s]"
          }
        },
        "d970095f0b9a477aa20b54b0c28b4f41": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "3db97d36de3a4f1e989350a37c6a286e": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "54954ee923594839b674c422e552fcce": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "6e9fb008b2ff4ed0892882dc8aa78f88": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "a7ac8db34f9340ea994ed66f8823a300": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "624ecab9cbe64c3e8c8b2ed96733d3fa": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d8f5c0d33a8c4d5f91164a54236e3e47": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/alfondace/ISYS5002/blob/main/08_text_sumarizier.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Text Summarisation\n",
        "\n",
        "There’s sooo much content to take in these days. \n",
        "\n",
        "* Blog posts\n",
        "* YouTube\n",
        "* Podcasts\n",
        "* Reports\n",
        "\n",
        "Wouldn't it be great to be able to summarise \n",
        "\n",
        "Using Hugging Face Transformers you can leverage a pre-trained summarisation pipeline to start summarising content.\n",
        "\n",
        "In this notebook you'll go through: \n",
        "1. Installing Hugging Face Transformers\n",
        "2. Building a summarization pipeline\n",
        "3. Running an encoding decoding model for summarization\n",
        "\n",
        "\n",
        "# Task\n",
        "\n",
        "We will write a notebook to convert a PDF or Word document to text or Web pages to text. The strategy is to convert everything to text and then summarise the text.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "SmVEXRTA1wZn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Summarizing text\n",
        "1. Installing Hugging Face Transformers\n",
        "2. Building a summarization pipeline\n",
        "3. Run model/pipeline to summarisation\n",
        "4. [Hugging Face Transformers](https://huggingface.co/docs/transformers/index) free state-of-the-art pre-trained machine learning models for processing text, images, audio and video. See the project website for more information.\n",
        "\n",
        "AI Text Summarization with Hugging Face Transformers in 4 Lines of Python: \n",
        "https://www.youtube.com/watch?v=TsfLm5iiYb4"
      ],
      "metadata": {
        "id": "FHtmW79E_jJS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Install Hugging Face Transformers and Import Dependencies"
      ],
      "metadata": {
        "id": "VA5FyTIOmSyD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install transformers"
      ],
      "metadata": {
        "id": "6VaWKTfymjoQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "761e41da-a8ff-40d0-f0ad-b28e8ad9824c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting transformers\n",
            "  Downloading transformers-4.22.1-py3-none-any.whl (4.9 MB)\n",
            "\u001b[K     |████████████████████████████████| 4.9 MB 17.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: tqdm>=4.27 in /usr/local/lib/python3.7/dist-packages (from transformers) (4.64.1)\n",
            "Collecting tokenizers!=0.11.3,<0.13,>=0.11.1\n",
            "  Downloading tokenizers-0.12.1-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (6.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.6 MB 60.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.7/dist-packages (from transformers) (6.0)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from transformers) (3.8.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from transformers) (2.23.0)\n",
            "Requirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.7/dist-packages (from transformers) (1.21.6)\n",
            "Collecting huggingface-hub<1.0,>=0.9.0\n",
            "  Downloading huggingface_hub-0.9.1-py3-none-any.whl (120 kB)\n",
            "\u001b[K     |████████████████████████████████| 120 kB 81.0 MB/s \n",
            "\u001b[?25hRequirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from transformers) (21.3)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.7/dist-packages (from transformers) (2022.6.2)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.7/dist-packages (from transformers) (4.12.0)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in /usr/local/lib/python3.7/dist-packages (from huggingface-hub<1.0,>=0.9.0->transformers) (4.1.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->transformers) (3.0.9)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata->transformers) (3.8.1)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (2.10)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (2022.6.15)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (1.24.3)\n",
            "Installing collected packages: tokenizers, huggingface-hub, transformers\n",
            "Successfully installed huggingface-hub-0.9.1 tokenizers-0.12.1 transformers-4.22.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#import dependecies\n",
        "from transformers import pipeline"
      ],
      "metadata": {
        "id": "w5by8iL6n3Ze",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 104,
          "referenced_widgets": [
            "23618c5d35fa4d42938a8037c995fd14",
            "7c25053baff7404bbcab92364d61cd05",
            "aa62410d6737451187ff528af39a37e9",
            "f7cb75813e874fa69f81be93c3fac28d",
            "12f97800aacd4de7af630f64b535db7b",
            "856434b2de6d4f749895c07156094209",
            "bfffee1ab7d54907bb4b962a2d652afb",
            "febd2c6447af46f1bb1aeb0c53efafc0",
            "210407cf4be94e99a15b650723167f0d",
            "74851daed77f493d81cadd73f3969b42",
            "9029b80b30c74a6c9bed95baa5545b20"
          ]
        },
        "outputId": "525759b0-39a1-4632-e27a-78fe93b091ba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Moving 0 files to the new cache system\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "0it [00:00, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "23618c5d35fa4d42938a8037c995fd14"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Build a pipeline and pass through text\n",
        "Load the pre-trained 'Summarization' from Pipeline"
      ],
      "metadata": {
        "id": "Aq6uBBitmaXt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "summarizer = pipeline('summarization')"
      ],
      "metadata": {
        "id": "4EYAYJqhmkQN",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 212,
          "referenced_widgets": [
            "80f0b7d8bfcb40f78ce491ed227761e8",
            "904776d93faa44958ba68487e87251e7",
            "723b413007b242918506390013ee0541",
            "7f0ee27686fa4c88b866d37893777978",
            "0de26fcbd73e4eadab35eb21c3faefb3",
            "0cb06b65aff14f38907b4f06e5d362be",
            "4f5bc8679cc7447ea3932e37fd92e7a2",
            "0c82f3ff048e40c1838ec1f25e382d12",
            "b74502b83e5149b385c60dfd6ec00d4a",
            "65fb1c906fc04687b460a5fe5b78a666",
            "b838d14e85fa4ebdbfd88a662c0a7540",
            "f84f090c0dec465c8a9490e1f50e7e74",
            "1c20e9f6f07046d89dac6a186402fb7c",
            "f4c9c81440f3467587dfab9dac2cf9eb",
            "26d4d2b134ec4a5baf3571668fc2bf84",
            "c8ad5be4d12b4cdd881fe0e17d5f78e1",
            "174725c73a9f4f2e87b758772a5936af",
            "a54813923ec04ed0babc76f192d40027",
            "fa48ac996102414b909d6fed0a28912f",
            "eb2194b21480456eb72b2edc3d35568e",
            "5733a9722b24405ba258cba1b7de50a1",
            "0b76ad47a8c84771b855b4d9fa780adb",
            "38fdfbc7ae7b4fc096345eb2bbb4fef9",
            "7c162c141fdf4330bec85b4e2c50bbbf",
            "0faf0b13b1174d5086953cb85d7bcb23",
            "eb506ea11331493886ae62a45342960b",
            "ad2baddab175472c8d0df25bbddafdd7",
            "25ab995d3fea419f8bcd024363d87fbb",
            "d7375a7afe6f406a98c414641a9f8abe",
            "bef5a9799baa4127b5afa9c2c0b97f70",
            "f99af7c7746d4a0bb4d9576a27943c57",
            "35684ce6f4d54c33800041aef11e9203",
            "dea024743e2c4f1b9509481411522940",
            "e58939836cd24adaade2f6b77c2260f8",
            "7268b6912e8a4de39c0deb7955649262",
            "5e39735d65c8460aad3ade749330320e",
            "97df69b0220646b7957d9c0fbd94d2f2",
            "030c740e71894835a2a1e6dd60caa853",
            "d84a371009b249cb922ab5d4e8688718",
            "89eeb246e8c94af7a665d2a1a2fbd326",
            "67e201ede8414ba98731a481ed4c6128",
            "6257c8fc4411421fba4d38f5dc015b42",
            "9d15cb0f3a1e4e09970408bccda8a979",
            "7aed8bb753bb487e8fa951b147804921",
            "a5e856ba81ce47edaf5fcce079b75aa5",
            "398a521c850341e390f35299c68ca316",
            "05987e78a5764fb99e31fed747ba1a86",
            "dd72772cd6ca4f5a8c3167fc933cafca",
            "d970095f0b9a477aa20b54b0c28b4f41",
            "3db97d36de3a4f1e989350a37c6a286e",
            "54954ee923594839b674c422e552fcce",
            "6e9fb008b2ff4ed0892882dc8aa78f88",
            "a7ac8db34f9340ea994ed66f8823a300",
            "624ecab9cbe64c3e8c8b2ed96733d3fa",
            "d8f5c0d33a8c4d5f91164a54236e3e47"
          ]
        },
        "outputId": "ee083b41-e622-4e95-ee51-4a148989480a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).\n",
            "Using a pipeline without specifying a model name and revision in production is not recommended.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Downloading:   0%|          | 0.00/1.80k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "80f0b7d8bfcb40f78ce491ed227761e8"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Downloading:   0%|          | 0.00/1.22G [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "f84f090c0dec465c8a9490e1f50e7e74"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Downloading:   0%|          | 0.00/26.0 [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "38fdfbc7ae7b4fc096345eb2bbb4fef9"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Downloading:   0%|          | 0.00/899k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "e58939836cd24adaade2f6b77c2260f8"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Downloading:   0%|          | 0.00/456k [00:00<?, ?B/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "a5e856ba81ce47edaf5fcce079b75aa5"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "help(summarizer)"
      ],
      "metadata": {
        "id": "3h89s64frjgp",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "3578bd1d-e31d-4e44-b78f-d5c945d330dd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on SummarizationPipeline in module transformers.pipelines.text2text_generation object:\n",
            "\n",
            "class SummarizationPipeline(Text2TextGenerationPipeline)\n",
            " |  SummarizationPipeline(*args, **kwargs)\n",
            " |  \n",
            " |  Summarize news articles and other documents.\n",
            " |  \n",
            " |  This summarizing pipeline can currently be loaded from [`pipeline`] using the following task identifier:\n",
            " |  `\"summarization\"`.\n",
            " |  \n",
            " |  The models that this pipeline can use are models that have been fine-tuned on a summarization task, which is\n",
            " |  currently, '*bart-large-cnn*', '*t5-small*', '*t5-base*', '*t5-large*', '*t5-3b*', '*t5-11b*'. See the up-to-date\n",
            " |  list of available models on [huggingface.co/models](https://huggingface.co/models?filter=summarization).\n",
            " |  \n",
            " |  Usage:\n",
            " |  \n",
            " |  ```python\n",
            " |  # use bart in pytorch\n",
            " |  summarizer = pipeline(\"summarization\")\n",
            " |  summarizer(\"An apple a day, keeps the doctor away\", min_length=5, max_length=20)\n",
            " |  \n",
            " |  # use t5 in tf\n",
            " |  summarizer = pipeline(\"summarization\", model=\"t5-base\", tokenizer=\"t5-base\", framework=\"tf\")\n",
            " |  summarizer(\"An apple a day, keeps the doctor away\", min_length=5, max_length=20)\n",
            " |  ```\n",
            " |  Arguments:\n",
            " |      model ([`PreTrainedModel`] or [`TFPreTrainedModel`]):\n",
            " |          The model that will be used by the pipeline to make predictions. This needs to be a model inheriting from\n",
            " |          [`PreTrainedModel`] for PyTorch and [`TFPreTrainedModel`] for TensorFlow.\n",
            " |      tokenizer ([`PreTrainedTokenizer`]):\n",
            " |          The tokenizer that will be used by the pipeline to encode data for the model. This object inherits from\n",
            " |          [`PreTrainedTokenizer`].\n",
            " |      modelcard (`str` or [`ModelCard`], *optional*):\n",
            " |          Model card attributed to the model for this pipeline.\n",
            " |      framework (`str`, *optional*):\n",
            " |          The framework to use, either `\"pt\"` for PyTorch or `\"tf\"` for TensorFlow. The specified framework must be\n",
            " |          installed.\n",
            " |  \n",
            " |          If no framework is specified, will default to the one currently installed. If no framework is specified and\n",
            " |          both frameworks are installed, will default to the framework of the `model`, or to PyTorch if no model is\n",
            " |          provided.\n",
            " |      task (`str`, defaults to `\"\"`):\n",
            " |          A task-identifier for the pipeline.\n",
            " |      num_workers (`int`, *optional*, defaults to 8):\n",
            " |          When the pipeline will use *DataLoader* (when passing a dataset, on GPU for a Pytorch model), the number of\n",
            " |          workers to be used.\n",
            " |      batch_size (`int`, *optional*, defaults to 1):\n",
            " |          When the pipeline will use *DataLoader* (when passing a dataset, on GPU for a Pytorch model), the size of\n",
            " |          the batch to use, for inference this is not always beneficial, please read [Batching with\n",
            " |          pipelines](https://huggingface.co/transformers/main_classes/pipelines.html#pipeline-batching) .\n",
            " |      args_parser ([`~pipelines.ArgumentHandler`], *optional*):\n",
            " |          Reference to the object in charge of parsing supplied pipeline parameters.\n",
            " |      device (`int`, *optional*, defaults to -1):\n",
            " |          Device ordinal for CPU/GPU supports. Setting this to -1 will leverage CPU, a positive will run the model on\n",
            " |          the associated CUDA device id. You can pass native `torch.device` or a `str` too.\n",
            " |      binary_output (`bool`, *optional*, defaults to `False`):\n",
            " |          Flag indicating if the output the pipeline should happen in a binary format (i.e., pickle) or as raw text.\n",
            " |  \n",
            " |  Method resolution order:\n",
            " |      SummarizationPipeline\n",
            " |      Text2TextGenerationPipeline\n",
            " |      transformers.pipelines.base.Pipeline\n",
            " |      transformers.pipelines.base._ScikitCompat\n",
            " |      abc.ABC\n",
            " |      builtins.object\n",
            " |  \n",
            " |  Methods defined here:\n",
            " |  \n",
            " |  __call__(self, *args, **kwargs)\n",
            " |      Summarize the text(s) given as inputs.\n",
            " |      \n",
            " |      Args:\n",
            " |          documents (*str* or `List[str]`):\n",
            " |              One or several articles (or one list of articles) to summarize.\n",
            " |          return_text (`bool`, *optional*, defaults to `True`):\n",
            " |              Whether or not to include the decoded texts in the outputs\n",
            " |          return_tensors (`bool`, *optional*, defaults to `False`):\n",
            " |              Whether or not to include the tensors of predictions (as token indices) in the outputs.\n",
            " |          clean_up_tokenization_spaces (`bool`, *optional*, defaults to `False`):\n",
            " |              Whether or not to clean up the potential extra spaces in the text output.\n",
            " |          generate_kwargs:\n",
            " |              Additional keyword arguments to pass along to the generate method of the model (see the generate method\n",
            " |              corresponding to your framework [here](./model#generative-models)).\n",
            " |      \n",
            " |      Return:\n",
            " |          A list or a list of list of `dict`: Each result comes as a dictionary with the following keys:\n",
            " |      \n",
            " |          - **summary_text** (`str`, present when `return_text=True`) -- The summary of the corresponding input.\n",
            " |          - **summary_token_ids** (`torch.Tensor` or `tf.Tensor`, present when `return_tensors=True`) -- The token\n",
            " |            ids of the summary.\n",
            " |  \n",
            " |  check_inputs(self, input_length: int, min_length: int, max_length: int) -> bool\n",
            " |      Checks whether there might be something wrong with given input with regard to the model.\n",
            " |  \n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data and other attributes defined here:\n",
            " |  \n",
            " |  __abstractmethods__ = frozenset()\n",
            " |  \n",
            " |  return_name = 'summary'\n",
            " |  \n",
            " |  ----------------------------------------------------------------------\n",
            " |  Methods inherited from Text2TextGenerationPipeline:\n",
            " |  \n",
            " |  __init__(self, *args, **kwargs)\n",
            " |      Initialize self.  See help(type(self)) for accurate signature.\n",
            " |  \n",
            " |  postprocess(self, model_outputs, return_type=<ReturnType.TEXT: 1>, clean_up_tokenization_spaces=False)\n",
            " |      Postprocess will receive the raw outputs of the `_forward` method, generally tensors, and reformat them into\n",
            " |      something more friendly. Generally it will output a list or a dict or results (containing just strings and\n",
            " |      numbers).\n",
            " |  \n",
            " |  preprocess(self, inputs, truncation=<TruncationStrategy.DO_NOT_TRUNCATE: 'do_not_truncate'>, **kwargs)\n",
            " |      Preprocess will take the `input_` of a specific pipeline and return a dictionnary of everything necessary for\n",
            " |      `_forward` to run properly. It should contain at least one tensor, but might have arbitrary other items.\n",
            " |  \n",
            " |  ----------------------------------------------------------------------\n",
            " |  Methods inherited from transformers.pipelines.base.Pipeline:\n",
            " |  \n",
            " |  check_model_type(self, supported_models: Union[List[str], dict])\n",
            " |      Check if the model class is in supported by the pipeline.\n",
            " |      \n",
            " |      Args:\n",
            " |          supported_models (`List[str]` or `dict`):\n",
            " |              The list of models supported by the pipeline, or a dictionary with model class values.\n",
            " |  \n",
            " |  device_placement(self)\n",
            " |      Context Manager allowing tensor allocation on the user-specified device in framework agnostic way.\n",
            " |      \n",
            " |      Returns:\n",
            " |          Context manager\n",
            " |      \n",
            " |      Examples:\n",
            " |      \n",
            " |      ```python\n",
            " |      # Explicitly ask for tensor allocation on CUDA device :0\n",
            " |      pipe = pipeline(..., device=0)\n",
            " |      with pipe.device_placement():\n",
            " |          # Every framework specific tensor allocation will be done on the request device\n",
            " |          output = pipe(...)\n",
            " |      ```\n",
            " |  \n",
            " |  ensure_tensor_on_device(self, **inputs)\n",
            " |      Ensure PyTorch tensors are on the specified device.\n",
            " |      \n",
            " |      Args:\n",
            " |          inputs (keyword arguments that should be `torch.Tensor`, the rest is ignored):\n",
            " |              The tensors to place on `self.device`.\n",
            " |          Recursive on lists **only**.\n",
            " |      \n",
            " |      Return:\n",
            " |          `Dict[str, torch.Tensor]`: The same as `inputs` but on the proper device.\n",
            " |  \n",
            " |  forward(self, model_inputs, **forward_params)\n",
            " |  \n",
            " |  get_inference_context(self)\n",
            " |  \n",
            " |  get_iterator(self, inputs, num_workers: int, batch_size: int, preprocess_params, forward_params, postprocess_params)\n",
            " |  \n",
            " |  iterate(self, inputs, preprocess_params, forward_params, postprocess_params)\n",
            " |  \n",
            " |  predict(self, X)\n",
            " |      Scikit / Keras interface to transformers' pipelines. This method will forward to __call__().\n",
            " |  \n",
            " |  run_multi(self, inputs, preprocess_params, forward_params, postprocess_params)\n",
            " |  \n",
            " |  run_single(self, inputs, preprocess_params, forward_params, postprocess_params)\n",
            " |  \n",
            " |  save_pretrained(self, save_directory: str)\n",
            " |      Save the pipeline's model and tokenizer.\n",
            " |      \n",
            " |      Args:\n",
            " |          save_directory (`str`):\n",
            " |              A path to the directory where to saved. It will be created if it doesn't exist.\n",
            " |  \n",
            " |  transform(self, X)\n",
            " |      Scikit / Keras interface to transformers' pipelines. This method will forward to __call__().\n",
            " |  \n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data and other attributes inherited from transformers.pipelines.base.Pipeline:\n",
            " |  \n",
            " |  default_input_names = None\n",
            " |  \n",
            " |  ----------------------------------------------------------------------\n",
            " |  Data descriptors inherited from transformers.pipelines.base._ScikitCompat:\n",
            " |  \n",
            " |  __dict__\n",
            " |      dictionary for instance variables (if defined)\n",
            " |  \n",
            " |  __weakref__\n",
            " |      list of weak references to the object (if defined)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Summarize Text"
      ],
      "metadata": {
        "id": "pqP63GnEmf0T"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "article = \"\"\"\n",
        "Around the world, as regulators look to rein in Big Tech, like the ongoing digital platforms inquiry in Australia, online platforms will face a raft of new rules in the EU.\n",
        "Known as the Digital Services Act, it’s a comprehensive set of regulations for digital services and content in the Eurozone.\n",
        "Like GDPR, the Digital Services Act is expected to lead the way for other countries to provide some rules around how digital services function, \n",
        "with everything from algorithms to online marketplaces, social networks, content-sharing platforms, app stores and online travel and accommodation platforms included.\n",
        "The Digital Services Act sets out clear due diligence obligations for digital platforms and other online intermediaries with measures for cooperation with trusted flaggers and \n",
        "competent authorities on content moderation, and measures to deter rogue traders from reaching consumers.\n",
        "\"\"\""
      ],
      "metadata": {
        "id": "48RNuGBQIHQ0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "summarizer(article, min_length=25, max_length=100)"
      ],
      "metadata": {
        "id": "PV4XVv2r-_fA",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "21c4602f-7d23-4e15-e574-797be6610c91"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[{'summary_text': ' The Digital Services Act is a comprehensive set of regulations for digital services and content in the Eurozone . It is expected to lead the way for other countries to provide some rules around how digital services function .'}]"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "summary = summarizer(article, min_length=25, max_length=100)"
      ],
      "metadata": {
        "id": "ipv7UdqRcmBQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "summary[0]['summary_text']"
      ],
      "metadata": {
        "id": "DYpTLpKEcqFA",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 53
        },
        "outputId": "ed1f5ebd-247d-4985-e732-c9ce70dd9551"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "' The Digital Services Act is a comprehensive set of regulations for digital services and content in the Eurozone . It is expected to lead the way for other countries to provide some rules around how digital services function .'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "summary[0]['summary_text'].split('.')"
      ],
      "metadata": {
        "id": "qYr5o3C0dBF4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "520b867e-a810-4c12-da86-f213bdb03336"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[' The Digital Services Act is a comprehensive set of regulations for digital services and content in the Eurozone ',\n",
              " ' It is expected to lead the way for other countries to provide some rules around how digital services function ',\n",
              " '']"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "adHCzKQWt00i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Summarise PDF using PyMuPDF"
      ],
      "metadata": {
        "id": "c8wd_jI48SHf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install wget"
      ],
      "metadata": {
        "id": "6ijHMepx__g6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a4e55cb9-2e19-47aa-965b-c2ed84514769"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting wget\n",
            "  Downloading wget-3.2.zip (10 kB)\n",
            "Building wheels for collected packages: wget\n",
            "  Building wheel for wget (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9675 sha256=6e12733a33038d48bc26e76e665ce2b3292f17d4c390480b930087706c7d03ec\n",
            "  Stored in directory: /root/.cache/pip/wheels/a1/b6/7c/0e63e34eb06634181c63adacca38b79ff8f35c37e3c13e3c02\n",
            "Successfully built wget\n",
            "Installing collected packages: wget\n",
            "Successfully installed wget-3.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Download a pdf file"
      ],
      "metadata": {
        "id": "0DHBSr_AsX0K"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Google Scholar 'Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud computing'\n",
        "!wget https://link.springer.com/content/pdf/10.1007/s11306-019-1588-0.pdf"
      ],
      "metadata": {
        "id": "Q22ZQqr6Q7Ti",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c0274989-2491-4e9f-a2fc-2a0f350189b8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-09-19 05:25:56--  https://link.springer.com/content/pdf/10.1007/s11306-019-1588-0.pdf\n",
            "Resolving link.springer.com (link.springer.com)... 151.101.0.95, 151.101.64.95, 151.101.128.95, ...\n",
            "Connecting to link.springer.com (link.springer.com)|151.101.0.95|:443... connected.\n",
            "HTTP request sent, awaiting response... 303 See Other\n",
            "Location: https://idp.springer.com/authorize?response_type=cookie&client_id=springerlink&redirect_uri=https%3A%2F%2Flink.springer.com%2Fcontent%2Fpdf%2F10.1007%2Fs11306-019-1588-0.pdf [following]\n",
            "--2022-09-19 05:25:56--  https://idp.springer.com/authorize?response_type=cookie&client_id=springerlink&redirect_uri=https%3A%2F%2Flink.springer.com%2Fcontent%2Fpdf%2F10.1007%2Fs11306-019-1588-0.pdf\n",
            "Resolving idp.springer.com (idp.springer.com)... 151.101.0.95, 151.101.64.95, 151.101.128.95, ...\n",
            "Connecting to idp.springer.com (idp.springer.com)|151.101.0.95|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://link.springer.com/content/pdf/10.1007/s11306-019-1588-0.pdf [following]\n",
            "--2022-09-19 05:25:56--  https://link.springer.com/content/pdf/10.1007/s11306-019-1588-0.pdf\n",
            "Connecting to link.springer.com (link.springer.com)|151.101.0.95|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1618179 (1.5M) [application/pdf]\n",
            "Saving to: ‘s11306-019-1588-0.pdf’\n",
            "\n",
            "s11306-019-1588-0.p 100%[===================>]   1.54M  --.-KB/s    in 0.01s   \n",
            "\n",
            "2022-09-19 05:25:57 (150 MB/s) - ‘s11306-019-1588-0.pdf’ saved [1618179/1618179]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "OR"
      ],
      "metadata": {
        "id": "aHEHtijvsUgJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import wget\n",
        "\n",
        "site_url = 'https://link.springer.com/content/pdf/10.1007/s11306-019-1588-0.pdf'\n",
        "file_name = wget.download(site_url)\n",
        "print(file_name)"
      ],
      "metadata": {
        "id": "EkJ28SYuMclv",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a7315773-4844-4392-f5a7-bf2b29b1eb00"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "s11306-019-1588-0 (1).pdf\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install PyMuPDF"
      ],
      "metadata": {
        "id": "msDeWTineRGf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0438c2e0-488e-4504-8b68-e97855473208"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: PyMuPDF in /usr/local/lib/python3.7/dist-packages (1.20.2)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import fitz # this is Pymupdf\n",
        "\n",
        "with fitz.open(file_name) as doc:\n",
        "  text = \"\"\n",
        "  for page in doc:\n",
        "    text = text + page.get_text()\n",
        "    \n",
        "print(text)"
      ],
      "metadata": {
        "id": "qaXAc_SQeZnp",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "673a6145-a595-4021-85d8-618cbc431097"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Vol.:(0123456789)\n",
            "1 3\n",
            "Metabolomics (2019) 15:125 \n",
            "https://doi.org/10.1007/s11306-019-1588-0\n",
            "REVIEW ARTICLE\n",
            "Toward collaborative open data science in metabolomics using \n",
            "Jupyter Notebooks and cloud computing\n",
            "Kevin M. Mendez1  · Leighton Pritchard2  · Stacey N. Reinke1  · David I. Broadhurst1 \n",
            "Received: 30 May 2019 / Accepted: 7 September 2019 / Published online: 14 September 2019 \n",
            "© The Author(s) 2019\n",
            "Abstract\n",
            "Background A lack of transparency and reporting standards in the scientific community has led to increasing and widespread \n",
            "concerns relating to reproduction and integrity of results. As an omics science, which generates vast amounts of data and \n",
            "relies heavily on data science for deriving biological meaning, metabolomics is highly vulnerable to irreproducibility. The \n",
            "metabolomics community has made substantial efforts to align with FAIR data standards by promoting open data formats, \n",
            "data repositories, online spectral libraries, and metabolite databases. Open data analysis platforms also exist; however, \n",
            "they tend to be inflexible and rely on the user to adequately report their methods and results. To enable FAIR data science \n",
            "in metabolomics, methods and results need to be transparently disseminated in a manner that is rapid, reusable, and fully \n",
            "integrated with the published work. To ensure broad use within the community such a framework also needs to be inclusive \n",
            "and intuitive for both computational novices and experts alike.\n",
            "Aim of Review To encourage metabolomics researchers from all backgrounds to take control of their own data science, mould \n",
            "it to their personal requirements, and enthusiastically share resources through open science.\n",
            "Key Scientific Concepts of Review This tutorial introduces the concept of interactive web-based computational laboratory \n",
            "notebooks. The reader is guided through a set of experiential tutorials specifically targeted at metabolomics researchers, \n",
            "based around the Jupyter Notebook web application, GitHub data repository, and Binder cloud computing platform.\n",
            "Keywords Open access · Reproducibility · Data science · Statistics · Cloud computing · Jupyter\n",
            "1 Introduction\n",
            "Historically, journal articles have been the primary medium \n",
            "for sharing new scientific research. The intent of article \n",
            "content, and the corresponding review process, is to ensure \n",
            "adequate evidence of reproducibility; however, a recent \n",
            "report highlights increasing and widespread concerns relat-\n",
            "ing to reproduction and integrity of results, with 52% of \n",
            "responding scientists agreeing there is a significant ‘crisis’ \n",
            "of reproducibility (Baker 2016). We and many others in the \n",
            "metabolomics community hold the view that a lack of trans-\n",
            "parency and incomplete reporting has led to significant mis-\n",
            "interpretation of data and a lack of trust in reported results \n",
            "(Broadhurst and Kell 2006; Considine et al. 2017; Goodacre \n",
            "et al. 2007; Spicer et al. 2017; Xia et al. 2013). A mechanism \n",
            "that may address these concerns is for the scientific com-\n",
            "munity to take advantage of new online publishing media \n",
            "and associated data services, encouraging open science that \n",
            "recognises and aligns with the FAIR (Findable, Accessible, \n",
            "Kevin M. Mendez and Leighton Pritchard have contributed equally \n",
            "to this article.\n",
            "Electronic supplementary material The online version of this \n",
            "article (https ://doi.org/10.1007/s1130 6-019-1588-0) contains \n",
            "supplementary material, which is available to authorized users.\n",
            " * Stacey N. Reinke \n",
            " \n",
            "stacey.n.reinke@ecu.edu.au\n",
            " * David I. Broadhurst \n",
            " \n",
            "d.broadhurst@ecu.edu.au\n",
            "1 \n",
            "Centre for Metabolomics & Computational Biology, School \n",
            "of Science, Edith Cowan University, Joondalup 6027, \n",
            "Australia\n",
            "2 \n",
            "Strathclyde Institute of Pharmacy & Biomedical \n",
            "Sciences, University of Strathclyde, Cathedral Street, \n",
            "Glasgow G1 1XQ, Scotland, UK\n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 2 of 16\n",
            "Interoperable, and Reusable) data principles (Wilkinson \n",
            "et al. 2016).\n",
            "This concern in metabolomics and other post-genomic \n",
            "platforms is a consequence of their success. The unprec-\n",
            "edented rate at which new mathematical algorithms and \n",
            "computational tools are developed and adopted means that \n",
            "published findings are increasingly the sole result of com-\n",
            "putationally intensive data processing (Teschendorff 2019). \n",
            "Advances in measurement technologies continue to gener-\n",
            "ate ever increasing volumes of high-throughput data, which \n",
            "in turn require multidisciplinary expertise as ever more \n",
            "complex and elaborate statistical methods are used to infer \n",
            "generalisable biological associations (Pinu et al. 2019) and \n",
            "sophisticated visualization tools are used to make large data-\n",
            "sets more understandable (Gehlenborg et al. 2010; Holten \n",
            "2006). Indeed, recent advances in machine learning algo-\n",
            "rithms combined with improved visualisation strategies now \n",
            "allow researchers to integrate and interrogate multi-omic \n",
            "data sets of a size that was unmanageable 5 years ago (for \n",
            "example, Lee et al. 2019; Reinke et al. 2018; Rohart et al. \n",
            "2017). For science to continue to move forward under this \n",
            "deluge of data while avoiding technical debt and maintain-\n",
            "ing reproducibility, the processes surrounding research data \n",
            "management, storage, analysis, and presentation need to be \n",
            "agile, transparent, reusable, and recoverably attached to pub-\n",
            "lished work. A further challenge is that, to gain broad adop-\n",
            "tion and be used by busy practising researchers, frameworks \n",
            "conforming to these requirements must also be intuitive and \n",
            "accessible to users who may have limited computational \n",
            "expertise.\n",
            "Over the last several years, the metabolomics commu-\n",
            "nity has made bold strides towards adopting FAIR standards \n",
            "for data including: development of vendor-independent raw \n",
            "data formats (such as mzXML) (Pedrioli et al. 2004); open \n",
            "access data repositories such as MetaboLights (Haug et al. \n",
            "2012), and Metabolomics Workbench (Sud et al. 2016); open \n",
            "access online spectral reference libraries such as METLIN \n",
            "(Smith et al. 2005), mzCloud (https ://www.mzclo ud.org/), \n",
            "and MassBank (Horai et al. 2010); and online databases \n",
            "for metabolite identification and biochemical association \n",
            "such as HMDB (Wishart et al. 2018). These resources and \n",
            "others like them are fundamental to the future integrity of \n",
            "metabolomics as a science. It is well-recognised that open, \n",
            "interoperable datasets are essential for progress, and the \n",
            "computational tools and methods that convert, step-by-step, \n",
            "metabolite data to biochemical meaning also need to be \n",
            "FAIR (Wilkinson et al. 2016).\n",
            "Numerous groups within the metabolomics community \n",
            "actively work to standardise computational workflows and \n",
            "provide online tools for statistical analysis (some recent \n",
            "advances are discussed later in this paper). However, a \n",
            "common characteristic of many computational frameworks \n",
            "encountered by researchers is a tendency to be prescriptive, \n",
            "in that they provide a restricted set of well-curated “plug \n",
            "and play” computational stepping stones that enable only \n",
            "limited choices within the workflow framework. These \n",
            "constraints limit the ability of a user to fully exploit the \n",
            "provided methodologies, or to explore and develop new \n",
            "analytical approaches. Presentation of analysis steps as plug-\n",
            "gable “black box” approaches is convenient but diminishes \n",
            "opportunities for education and understanding of the analy-\n",
            "sis methods being used. To fully embrace the concept of \n",
            "‘open data science’ the metabolomics community needs an \n",
            "open and easily accessible computational environment for \n",
            "rapid collaboration and experimentation.\n",
            "The subject of this tutorial review is a practical open-sci-\n",
            "ence solution to this problem that balances ease-of-use and \n",
            "flexibility, specifically targeted to novice metabolomic data \n",
            "scientists. This solution takes the form of ‘computational lab \n",
            "books’, such as Jupyter Notebooks (Kluyver et al. 2016), that \n",
            "have a diverse range of overlapping potential applications \n",
            "in the post-genomic research community (Fig. 1). Firstly, \n",
            "they enable open collaboration by providing a central plat-\n",
            "form for researchers to cooperatively develop methodology \n",
            "and perform data analysis. Secondly, they provide a means \n",
            "for transparent dissemination of a finished study or product. \n",
            "In a formal context computational lab books can comprise \n",
            "supplemental material extending the reach of a publica-\n",
            "tion that enables readers to rapidly recreate data analyses \n",
            "Fig. 1  Applications for Jupyter Notebooks in the postgenomic com-\n",
            "munity. Open virtual notebooks have three main, non-mutually \n",
            "exclusive, applications. First, they provide an efficient means for \n",
            "transparent dissemination of methods and results, thereby enabling \n",
            "alignment with FAIR data principles. Second, they provide a central \n",
            "and interactive platform that facilitates open collaboration to develop \n",
            "methodology and perform data analysis. Finally, their interactive and \n",
            "easily deployable framework can drive experiential learning opportu-\n",
            "nities for computational novices to develop their own skills and better \n",
            "understand metabolomics data analysis\n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 3 of 16 \n",
            "125\n",
            "and figures for themselves. In an informal context, they can \n",
            "provide a polished “showcase” that allows users to interact \n",
            "with and understand the functionality of underlying algo-\n",
            "rithms. Finally, the inherent promotion of direct user interac-\n",
            "tion enables experiential learning opportunities, where the \n",
            "user develops their understanding and skills through active \n",
            "experimentation, reflective observation, and abstract con-\n",
            "ceptualisation (Kolb 1984).\n",
            "In this review, we provide a brief overview of current data \n",
            "science frameworks relevant to the metabolomics commu-\n",
            "nity, corresponding barriers to achieving open science, and \n",
            "finally a practical solution in the form of the computational \n",
            "lab notebook, where code, prose and figures are combined \n",
            "into an interactive notebook that can be published online \n",
            "and accessed in a modern web browser through cloud com-\n",
            "puting. We present a set of experiential learning tutorials \n",
            "introducing the Jupyter Notebook framework, specifically \n",
            "tailored to the needs of a metabolomics researcher. The \n",
            "tutorials are designed in a hierarchy of complexity follow-\n",
            "ing Bloom’s taxonomy of educational learning objectives \n",
            "(Anderson et al. 2001). Tutorial one introduces the basic \n",
            "concepts of Jupyter Notebooks. Tutorial two encourages \n",
            "interactive learning using an existing metabolomics data \n",
            "science Jupyter notebook. Tutorial three establishes the \n",
            "framework in which the user can create a Jupyter notebook \n",
            "on a local computer. Tutorial four teaches the user how to \n",
            "create a simple notebook for their own data. Tutorial five \n",
            "explains how to publish and share a new Jupyter notebook \n",
            "in the cloud. The overarching aim of this document is to \n",
            "encourage metabolomics researchers from all backgrounds, \n",
            "possibly with little or no computational expertise, to seize \n",
            "the opportunity to take control of their own data science, \n",
            "mould it to their personal requirements, and enthusiastically \n",
            "share resources through open science.\n",
            "2  Background\n",
            "A glossary of terms has been provided in Table 1 to help \n",
            "clarify technical terms used in this tutorial.\n",
            "2.1  Software tools and barriers to open science\n",
            "Many statistical and data science software tools are avail-\n",
            "able for use in metabolomics data analysis and visualisa-\n",
            "tion. They can be classified as commercial (requiring a paid \n",
            "licence) or “free” (as in zero-cost) and, in either case, may \n",
            "be open-source (the underlying computer code is available \n",
            "for inspection) or proprietary (closed-source, code unavail-\n",
            "able for inspection). The primary mode of interaction with \n",
            "the user may be via scripting, a command line (CLI), or a \n",
            "graphical (GUI) user interface.\n",
            "Commercial, proprietary (closed-source) GUI software \n",
            "packages include Microsoft Excel, Minitab, SPSS, Unscram-\n",
            "bler, and SIMCA (Umetrics). Tools like these generally \n",
            "offer the benefits of being user-friendly, stable and reliable \n",
            "platforms with well-documented resources, and have a high \n",
            "level of technical customer support. However, proprietary \n",
            "software can also lack methodological transparency because \n",
            "the source code is not freely available. When source code \n",
            "cannot be inspected a researcher’s ability to interrogate \n",
            "underlying algorithms, demonstrate correctness, diagnose \n",
            "problems, or improve the tool is limited. If the package \n",
            "prescribes an analytical workflow it may be difficult, or \n",
            "even impossible, to embed alternative third-party compu-\n",
            "tational steps. If additional functionality is required users \n",
            "are dependent on the software’s developers to implement \n",
            "this, which may impose an additional expensive commercial \n",
            "transaction even in cases where the request is approved. It is \n",
            "also difficult to produce usable graphical interfaces that are \n",
            "also customisable by the user, so this kind of interface can \n",
            "be relatively inflexible and so constrain the researcher to a \n",
            "specific mode of working.\n",
            "Command-line or script-based proprietary software pack-\n",
            "ages such as MATLAB, SAS, and Stata overcome some of \n",
            "the limitations imposed by graphical interfaces and closed-\n",
            "source code by allowing third party code to be embedded, \n",
            "and implementation of alternative algorithms and arbitrary \n",
            "workflows by the researcher. In the case of MATLAB the \n",
            "source code of some or all of the proprietary tools is read-\n",
            "able, which improves transparency of methods, and it is pos-\n",
            "sible for the programmer to develop open custom graphical \n",
            "interfaces. However, even then open-source commercial \n",
            "packages can carry a significant financial cost limiting the \n",
            "ability of researchers, especially those in developing nations \n",
            "or on smaller budgets, to replicate results, adapt methods, \n",
            "or collaborate to develop better workflows. We consider \n",
            "that open-source “free” tools and applications will form the \n",
            "future basis of shareable research, as they enable the greatest \n",
            "possible degree of transparency and reproducibility.\n",
            "Open-source GUI workflows providing simplified or user-\n",
            "friendly access to underlying programs and analytical tools \n",
            "have been developed to improve usability for scientists who \n",
            "have not yet acquired the programming skills necessary to \n",
            "write their own pipelines and workflows. Within the metab-\n",
            "olomics community popular applications include: Meta-\n",
            "boAnalyst (Xia and Wishart 2011), Galaxy-M (Davidson \n",
            "et al. 2016), and Workflow4Metabolomics (Giacomoni et al. \n",
            "2015). Galaxy workflows provide a unified data visualisa-\n",
            "tion and analysis environment that allows seamless (to the \n",
            "user) integration of multiple open-source software packages, \n",
            "and tools written in multiple programming languages (Afgan \n",
            "et al. 2018). These tools allow rapid construction, imple-\n",
            "mentation, and sharing of standardised workflows, includ-\n",
            "ing integration with remote and local databases, without the \n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 4 of 16\n",
            "need for programming skills. This provides a mechanism \n",
            "to ensure methodological consistency and precise report-\n",
            "ing standards. Resources such as Galaxy simplify the user \n",
            "experience and enable flexible use of a wide range of open \n",
            "source tools.\n",
            "Despite the many strengths of open-source GUI work-\n",
            "flows such as Galaxy, they do not always provide users \n",
            "with a free choice of available data analysis methods. For \n",
            "example, unless the user has administrative rights on the \n",
            "server, the browser interface of Galaxy does not permit \n",
            "direct access to software package management. This restricts \n",
            "extension, modification, and development of workflows by \n",
            "the user. Although an arbitrary set of tools can in principle \n",
            "be “wrapped” by a researcher for use with Galaxy, there may \n",
            "be in practice only limited support for requests to imple-\n",
            "ment a tool, especially when working on public servers. It \n",
            "is possible to implement arbitrary tools and processes in a \n",
            "locally-managed Galaxy instance with administrative control \n",
            "of the workflow service, but this requires investment of time, \n",
            "technical expertise, and local computational capacity, as well \n",
            "as carrying implications for long-term systems support and \n",
            "maintenance.\n",
            "Table 1  Glossary of terms\n",
            "Paper section Term\n",
            "Definition\n",
            "1\n",
            "Data repository\n",
            "A platform (such as Metabolights or Metabolomics Workbench) used to store meta-\n",
            "data and experimental data\n",
            "2.1\n",
            "Command line interface (CLI)\n",
            "A user interface that is used to execute operating system functions using text\n",
            "2.1\n",
            "Graphical user interface (GUI)\n",
            "A user interface that is used to execute operating system functions using graphical \n",
            "icons or other visual indicators\n",
            "2.1\n",
            "Integrated development environment (IDE) A software application that provides an interface to write and test code (such as \n",
            "RStudio, PyCharm and Visual Studio Code). It typically includes basic tools such \n",
            "as a code editor, compiler, and a debugger\n",
            "2.1\n",
            "Containers\n",
            "Self-contained units of software that package code, dependencies, system tools and \n",
            "system libraries. The purpose is to be reliably transferred between, and deployed \n",
            "on, various operating systems and infrastructures\n",
            "2.1\n",
            "JavaScript object notation (JSON) format\n",
            "A lightweight data-interchange format commonly used for communication between \n",
            "a browser and server. Internally, Jupyter Notebooks are JSON files with the.ipynb \n",
            "extension\n",
            "2.1\n",
            "Packages\n",
            "Units of shareable code that can be imported and used to provide additional func-\n",
            "tionality (such as matplotlib and scikit-learn)\n",
            "2.1\n",
            "Application programming interface (API)\n",
            "A set of defined functions and protocols for interacting with the software or package\n",
            "2.1\n",
            "Kernel\n",
            "The “computational engine” that runs and introspects the code contained in a note-\n",
            "book document. Jupyter supports a kernel for Python, as well as kernels for many \n",
            "other languages (such as R, Julia, Kotlin, etc.)\n",
            "2.2\n",
            "Version control\n",
            "A documented history of changes made to a file, enabling step-by-step reproduction \n",
            "and reconstruction of its development\n",
            "2.2\n",
            "Code repository\n",
            "A hosted archive (such as those at GitHub and BitBucket) of source code and sup-\n",
            "porting files.\n",
            "3\n",
            "Virtual environment\n",
            "An isolated environment that contains a specific version of Python and dependencies\n",
            "3.1.1\n",
            "Distribution (Software)\n",
            "A collection of software bundled together\n",
            "3.1.1\n",
            "Markdown\n",
            "A lightweight markup language used to add and format plain text. It is used in Jupy-\n",
            "ter Notebooks within “Markdown” cells\n",
            "3.1.3\n",
            "Configuration file\n",
            "A file used to set the initial settings and parameters for computer applications. It is \n",
            "used in Binder to build the virtual environment with specific dependencies\n",
            "3.2.1\n",
            "Text cell (Markdown cell)\n",
            "A cell in the Jupyter Notebook used to write text (using the Markdown language)\n",
            "3.2.1\n",
            "Code cell\n",
            "A cell in the Jupyter Notebook used to run code (such as Python code)\n",
            "3.2.3\n",
            "Sandbox (Software development)\n",
            "A software environment typically used to run or test experimental code in isolation \n",
            "from the rest of the system\n",
            "3.2.5\n",
            "Dependencies\n",
            "The packages (and versions) that are required to be installed to use the software. For \n",
            "Python, these are the packages that need to be imported at the start of the file\n",
            "3.2.5\n",
            "Channels (Specific to Anaconda)\n",
            "The location where packages that are installed using conda are stored (such as \n",
            "conda-forge and bioconda)\n",
            "3.2.5\n",
            "README\n",
            "A file (commonly markdown or text) used to communicate information to visitors \n",
            "about the repository (such as purpose, usage, and contributors)\n",
            "3.2.5\n",
            "Root directory\n",
            "The directory (or folder) that is the highest level in a hierarchy\n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 5 of 16 \n",
            "125\n",
            "Even with a free choice of tools and algorithms, work-\n",
            "flows implemented in GUI-based tools like Galaxy are “lin-\n",
            "ear” in the sense that the browser interface imposes a process \n",
            "in which data passes through a sequential chain of opera-\n",
            "tions. These interfaces are not well-suited to representing \n",
            "complex workflow logic, such as branches and loops that \n",
            "explore alternative approaches or parameter choices as part \n",
            "of the same analysis. This can inadvertently encourage a \n",
            "“black box” one-size-fits-all approach to analysis that may \n",
            "be of concern when the dataset is non-standard or, for exam-\n",
            "ple, when a statistical analysis requires customisation due to \n",
            "assumptions made by the model regarding the distribution \n",
            "of the input data. Incurious application of standardised GUI \n",
            "workflows with limited opportunity for experimentation can \n",
            "lead to inappropriate analytical strategies and unintentional \n",
            "misreporting of results. The linearity constraint is recog-\n",
            "nised by the Galaxy developers, who provide a program-\n",
            "matic Application Programming Interface (API) enabling \n",
            "automation of complex workflow logic, but this requires \n",
            "programming ability to use.\n",
            "Another limitation of GUI workflow-based applications \n",
            "can be a lack of contextual annotation. With most interfaces \n",
            "the user must document separately why computational \n",
            "methods and parameter settings were chosen in a specific \n",
            "workflow. It is not typically possible through GUI workflow \n",
            "interfaces to embed the experimental context, explanation \n",
            "of methods, code, and figures into a single live (interactive) \n",
            "document. The formal reporting may then be reduced to a \n",
            "terse listing of steps and parameter values for generating \n",
            "data, tables and figures, rather than a more readable “liter-\n",
            "ate programming” account of the analysis. This retrospec-\n",
            "tive approach is sufficient and appropriate for standardised, \n",
            "repeated workflows that vary little from experiment to exper-\n",
            "iment, such as a mass spectrometry deconvolution workflow \n",
            "that converts a set of raw instrument files into an annotated \n",
            "table (e.g. XCMS → CAMERA → MetFrag). However, when \n",
            "a metabolomics scientist moves on to statistical analysis, \n",
            "multivariate machine learning, and data visualisation to \n",
            "extract and present a biologically-informative interpreta-\n",
            "tion of the data, it is desirable to have an integrated, flexible \n",
            "data analysis environment that includes detailed annotation \n",
            "of analysis choices.\n",
            "The most flexible data science solution is to conduct \n",
            "analyses in one or more high-level open-source program-\n",
            "ming languages such as C, Fortran, Java, Julia, Perl, Python, \n",
            "Octave, R, or Scala, that also support sophisticated statistical \n",
            "tools. Python and R have become especially popular lan-\n",
            "guages in data science due to the availability of comprehen-\n",
            "sive, robust, and well-documented code libraries (modules/\n",
            "packages). Many statistical and machine learning packages \n",
            "are available for these languages (including bindings to Gal-\n",
            "axy, which overcomes some of the GUI-based limitations of \n",
            "that platform), with strong data science community support \n",
            "(Lantz 2013; Müller and Guido 2017). However, these gen-\n",
            "eral-purpose languages may present novice (or non) data \n",
            "scientists with a forbiddingly steep learning curve, especially \n",
            "in comparison with GUI tools. To be most effective in these \n",
            "languages a researcher requires a basic understanding of \n",
            "computer programming to use the available code libraries \n",
            "in their specific field. There is an initial learning curve, but \n",
            "knowledge of a programming language is more generally \n",
            "useful and broadly applicable than familiarity with a specific \n",
            "software tool’s interface and can impact positively on many \n",
            "areas of research. Programming is increasingly recognised \n",
            "as a foundational skill for research and promoted at all levels \n",
            "from primary to postgraduate education (Passey 2017). The \n",
            "broad impact of this skillset throughout academic research, \n",
            "including arts and humanities, is recognised in the growing \n",
            "influence of training foundations such as The Carpentries \n",
            "(https ://carpe ntrie s.org/) that aim to “[teach] researchers the \n",
            "computing skills they need to get more done in less time and \n",
            "with less pain.”\n",
            "Several freely-available software tools bridge the gap \n",
            "between GUI interfaces and high-level languages by pro-\n",
            "viding a user interface for researchers to develop their own \n",
            "code. For Python and R, integrated development environ-\n",
            "ments (IDEs) such as PyCharm (Python), RStudio (R), and \n",
            "more general multi-language IDEs (e.g. Visual Studio Code, \n",
            "Komodo and Eclipse), provide additional tools for automat-\n",
            "ing, testing and visualizing the process of writing scripts, \n",
            "programs and analysis workflows. These IDEs can simplify \n",
            "the learning and programming experience but are primarily \n",
            "designed for larger program and application development, \n",
            "rather than composing and sharing data analysis workflows. \n",
            "However, IDEs in general are extremely useful even to the \n",
            "novice programmer, and some prominent examples are spe-\n",
            "cifically targeted towards data analysis, such as RStudio and \n",
            "JupyterLab.\n",
            "Recently, several independent strands of general-purpose \n",
            "data science software development have been woven into \n",
            "practical solutions to the various limitations of the above \n",
            "frameworks. Firstly, RStudio established itself as the ‘go to’ \n",
            "data science IDE for R programming and was extended to \n",
            "allow integration of R code, narrative text, and figures into \n",
            "a single notebook interface using “RMarkdown” (Baumer \n",
            "et al. 2014). The software companies Enthought Inc. and \n",
            "Anaconda (formerly Continuum Analytics) independently \n",
            "developed distributions of the Python programming lan-\n",
            "guage to include core scientific computing packages. Ana-\n",
            "conda later extended their distribution to include R. In 2015, \n",
            "the non-profit Project Jupyter was established (Kluyver \n",
            "et al. 2016) to “develop open-source software, open-stand-\n",
            "ards, and services for interactive computing across dozens \n",
            "of programming languages” (Project Jupyter 2019). Their \n",
            "main product is Jupyter Notebook, a browser-based interac-\n",
            "tive data science notebook environment. Jupyter Notebook \n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 6 of 16\n",
            "allows seamless integration of code, narrative text, and fig-\n",
            "ures into a single live executable and editable document, \n",
            "recorded in the open-standard and language-independent \n",
            "JavaScript Object Notation (JSON) format. Notebooks may \n",
            "be written in a single programming language, or a combina-\n",
            "tion of multiple languages. Jupyter Notebooks can use ker-\n",
            "nels for new or more specialised languages (such as Kotlin, \n",
            "GAP, Haskell, etc.), which gives them an advantage of being \n",
            "agnostic to programming language. Finally, integration of \n",
            "Jupyter Notebooks with the Docker (www.docke r.com) vir-\n",
            "tualization platform enables operating system level work-\n",
            "ing environments to be packaged into virtual “containers”, \n",
            "which allows collections of notebooks and the supporting \n",
            "third-party tools and software to be deployed as public, self-\n",
            "contained, reproducible interactive services using cloud \n",
            "computing.\n",
            "2.2  Collaboration through cloud computing\n",
            "Open and dynamic collaboration on projects is critical to \n",
            "effective working but remains a significant challenge for \n",
            "researchers. There is a real and present need for efficient \n",
            "sharing and management of files that allows easy access, \n",
            "use, and version control (a documented history of the \n",
            "changes made to a file, enabling step-by-step reproduction \n",
            "and reconstruction of its development) for all collaborators. \n",
            "Widely-used collaboration mechanisms such as sharing code \n",
            "via email or online blogs are cumbersome, frequently lead-\n",
            "ing to conflicts between the work of different researchers \n",
            "as they work on the same files at the same time in different \n",
            "locations. Cloud services including Box, Google Drive, and \n",
            "Dropbox have become essential tools for scientists by pro-\n",
            "viding shared online data and document storage. Tools such \n",
            "as Microsoft Office Online and Google Suite provide real-\n",
            "time collaboration tools enabling true simultaneous editing \n",
            "of a single document by multiple authors, and services like \n",
            "Dropbox are able to track edits and prompt users to keep \n",
            "local copies of files up to date. Both approaches allow users \n",
            "to step back through document history as a rudimentary \n",
            "form of version control. They reduce practical barriers to \n",
            "collaborative working and reduce frustration and conflicts \n",
            "resulting from two or more people editing different copies \n",
            "of the same file at the same time. Collaborative working on \n",
            "metabolomic data analysis workflows would benefit from \n",
            "adoption of similar approaches.\n",
            "The source code hosting facilities Bitbucket, GitHub \n",
            "and SourceForge are currently the dominant platforms for \n",
            "sharing and collaborating on (particularly open-source) \n",
            "software. GitHub has become the largest source code host-\n",
            "ing facility in the world, with over 36 million users and \n",
            "100 million repositories (GitHub 2019). These facilities \n",
            "offer many benefits including: free public (and private) \n",
            "source code repositories; enforced best practice through \n",
            "version control; and additional administrative and project \n",
            "management services that foster collaboration, including \n",
            "project webpages and wikis, issue tracking, code reviews, \n",
            "and task management. This makes GitHub and similar \n",
            "services a practical option for development, publication \n",
            "and distribution of Jupyter Notebooks, together with their \n",
            "associated source code and test data.\n",
            "Services such as GitHub and BitBucket allow col-\n",
            "laborators to view and edit static code and view static \n",
            "notebooks, but code cannot be executed directly on their \n",
            "servers. To run Jupyter Notebooks and associated source \n",
            "code, the user must either download and run a local copy \n",
            "of the files, or upload and run the notebook “in the cloud” \n",
            "using a cloud infrastructure provider such as Amazon \n",
            "Web Services, Google Colab, Openstack, or Microsoft \n",
            "Azure. The process of enabling the practical use of this \n",
            "shared resource can therefore require a level of compu-\n",
            "tational expertise that may be a deterrent to casual users \n",
            "and restrict uptake by non-expert data-curious scientists.\n",
            "The PhenoMeNal portal (http://pheno menal -h2020 .eu) \n",
            "is an elegant solution to this problem for the metabolomics \n",
            "community. PhenoMeNal (Peters et al. 2019) is an easy-\n",
            "to-use, cloud-based metabolomics research environment \n",
            "led by EMBL’s European Bioinformatics Institute. The \n",
            "PhenoMeNal App Library includes over 50 widely used \n",
            "metabolomics data analysis tools that can be accessed \n",
            "either through Jupyter or Galaxy and deployed using a \n",
            "cloud infrastructure provider. This curated software library \n",
            "allows the community to maintain consistency across \n",
            "workflows but, in common with other GUI tools and cen-\n",
            "trally-managed workflow approaches, it can be restrictive.\n",
            "A comparable but completely general public service \n",
            "is provided by the Binder team at mybinder.org (Project \n",
            "Jupyter et al. 2018). Binder is an open-source web service \n",
            "that allows users to share notebooks by creating a tem-\n",
            "porary cloud-based copy of the GitHub repository that \n",
            "contains them. This enables reproducible sharing of inter-\n",
            "active and editable Jupyter or Rstudio notebooks as a vir-\n",
            "tual machine running in the cloud. The user can start and \n",
            "access a new virtual machine running live notebooks by \n",
            "following a single web link. In use, the notebooks appear \n",
            "to the user as if they were any other Jupyter notebook \n",
            "running on their own computer, with all the necessary \n",
            "dependencies, supplementary code and data pre-installed. \n",
            "Using the Binder framework gives researchers the power \n",
            "to reproduce and thoroughly test published results, or \n",
            "apply the analyses to their own data by running the source \n",
            "code interactively in their browser. In this tutorial review \n",
            "we take the reader through a process of using, writing, and \n",
            "deploying Jupyter Notebooks on Binder to help them take \n",
            "control of their own data science, and share their work \n",
            "through open science approaches.\n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 7 of 16 \n",
            "125\n",
            "3  Experiential learning tutorials\n",
            "The remainder of this review provides readers with an \n",
            "experiential learning opportunity (Kolb 1984) using an \n",
            "example interactive metabolomics data analysis workflow \n",
            "deployed using a combination of Python, Jupyter Note-\n",
            "books, and Binder. We assume that the initial stage of \n",
            "data-processing for the computational workflow (con-\n",
            "verting raw instrument files into an annotated data table) \n",
            "has already been completed, and that a deconvolved, but \n",
            "not necessarily annotated, data table has been created \n",
            "and checked for errors. These assumptions are made to \n",
            "make the learning objectives presented manageable, not \n",
            "as a directive for obfuscating the complete metabolomics \n",
            "workflow. It is possible, and encouraged, to include all \n",
            "data processing steps in interactive notebooks. The tutorial \n",
            "takes the reader through the process of using interactive \n",
            "notebooks to produce a shareable, reproducible data analy-\n",
            "sis workflow that connects the study design to reported \n",
            "biological conclusions in an interactive document, using \n",
            "data from two previously published metabolomics studies. \n",
            "This workflow includes a discrete set of interactive and \n",
            "interlinked procedures: data cleaning, univariate statistics, \n",
            "multivariate machine learning, feature selection, and data \n",
            "visualisation (Fig. 2).\n",
            "The following five tutorials have been pedagogically \n",
            "designed to lead the reader through increasing levels of \n",
            "cognitive complexity, according to Bloom’s revised tax-\n",
            "onomy (Anderson et al. 2001):\n",
            "(1) Launch and walk through a published Jupyter notebook \n",
            "using Binder in the cloud to duplicate a set of results.\n",
            "(2) Interact with and edit the content of a published Jupy-\n",
            "ter notebook using Binder in the cloud to understand \n",
            "workflow methods.\n",
            "(3) Install Python and use published Jupyter Notebooks on \n",
            "the researcher’s computer to apply and experiment with \n",
            "workflow methods locally.\n",
            "(4) Create a metabolomics Jupyter notebook on a local \n",
            "computer.\n",
            "(5) Deploy the Jupyter notebook from Tutorial 4 on Binder \n",
            "in the cloud via GitHub.\n",
            "Fig. 2  Metabolomics data analysis workflow. The workflow imple-\n",
            "mented in Tutorials 1 and 2 represents a typical metabolomics data \n",
            "science workflow for a binary classification outcome. The following \n",
            "steps are included: data import, data cleaning based on pooled QC \n",
            "relative standard deviation, PCA to visually inspect data reproduc-\n",
            "ibility, univariate statistics, multivariate machine learning (PLS-DA \n",
            "including cross validation, feature selection, and permutation test-\n",
            "ing). The flow diagram is coloured by primary operation type (yel-\n",
            "low = data import/export; green = data visualisation; blue = data pro-\n",
            "cessing)\n",
            "▸\n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 8 of 16\n",
            "3.1  Overview of Jupyter/GitHub/Binders\n",
            "Before beginning the tutorial, we review some fundamental \n",
            "concepts behind Jupyter Notebooks, GitHub, and Binder, \n",
            "as understanding these can aid successful independent \n",
            "execution of this open-science approach (Fig. 3). All code \n",
            "embedded in each of the example notebooks is written in \n",
            "the Python programming language and is based upon exten-\n",
            "sions of popular open source packages with high levels of \n",
            "community uptake and support. These include: Numpy for \n",
            "matrix-based calculations (van der Walt et al. 2011); Pandas \n",
            "for high level data table manipulation (McKinney 2017); \n",
            "Scikit-learn for machine learning (Pedregosa et al. 2011); \n",
            "and Matplotlib (Hunter 2007), Bokeh (Bokeh Development \n",
            "Team 2018), Seaborn (Waskom et al. 2018), and BeakerX \n",
            "(Beaker X Development Team 2018) for data visualisation. \n",
            "Additionally, we deploy a simple package called ‘cimcb-\n",
            "lite’, developed by the authors for this publication, that inte-\n",
            "grates the functionality of the above packages into a set of \n",
            "basic methods specific to metabolomics. A tutorial on the \n",
            "Python programming language itself is beyond the scope of \n",
            "this publication, but we hope that the code presented is suffi-\n",
            "ciently well-documented in each notebook to be understood. \n",
            "Many excellent publications can be consulted for an in-depth \n",
            "introduction to using Python for data science (Jones 2013; \n",
            "Ramalho 2015; The Carpentries 2019; VanderPlas 2016).\n",
            "Digital object identifiers (DOI) are widely used to iden-\n",
            "tify academic and government information in the form of \n",
            "journal articles, research reports and data sets. It is also pos-\n",
            "sible to assign a DOI to open access software. Specifically, \n",
            "researchers are able to make the work shared on GitHub cit-\n",
            "able by archiving with a data archiving tool such as Zenodo \n",
            "(www.zenod o.org) (Sicilia et al. 2017). A detailed tutorial is \n",
            "available (Open Science MOOC 2018). This archiving tool \n",
            "will ‘fix’ in time a given repository (e.g. Jupyter notebook \n",
            "and meta data), so that it can be associated with a particular \n",
            "static publication, while allowing the programmer to further \n",
            "develop the notebook on GitHub. The tutorials in this paper \n",
            "are archived with the handle https ://doi.org/10.5281/zenod \n",
            "o.33626 24 (https ://doi.org/10.5281/zenod o.33626 24).\n",
            "3.1.1  Jupyter Notebook\n",
            "Jupyter Notebook (jupyter.org) is a powerful, open-source, \n",
            "browser-based tool for interactive development and presen-\n",
            "tation of data science projects. Each notebook consists of a \n",
            "collection of executable cells, and each cell contains either \n",
            "text formatted using the Markdown language (Gruber 2004) \n",
            "or executable code (usually Python or R). When a ‘code \n",
            "cell’ is executed any graphical or text output (numerical \n",
            "results, figures or tables) is presented within the document \n",
            "immediately below the cell. Figure 4 shows an example of \n",
            "a notebook after execution. A popular way to get started \n",
            "with Jupyter Notebooks is to install the Anaconda distri-\n",
            "bution (anaco nda.com), for which graphical installers are \n",
            "available on Windows, macOS and Linux operating systems \n",
            "(anaco nda.com/distr ibuti on/). After installation a local Jupy-\n",
            "ter server can be launched using the Anaconda-Navigator \n",
            "application. To run a specific local Jupyter notebook with \n",
            "Anaconda-Navigator the user can navigate to the appropri-\n",
            "ate local folder using the browser-based interface, and click \n",
            "on the desired notebook file (which can be identified by the \n",
            ".ipynb suffix).\n",
            "3.1.2  GitHub\n",
            "GitHub (githu b.com) is a cloud-based web service that helps \n",
            "programmers store, manage, and share their code (and asso-\n",
            "ciated data files), as well as track and control changes to their \n",
            "code (version control). It is free to sign up and host a public \n",
            "code repository, which makes GitHub especially popular \n",
            "with open-source projects and a good choice for distributing \n",
            "Jupyter Notebooks, project-specific code and documenta-\n",
            "tion. Jupyter Notebooks stored publicly on GitHub can be \n",
            "downloaded and run on a local machine using Anaconda \n",
            "or linked to a cloud-based platform. To complete all the \n",
            "steps of this tutorial a (free) GitHub account is required. \n",
            "An account at GitHub may be created by clicking “sign up” \n",
            "on the GitHub home page (githu b.com) and following the \n",
            "instructions.\n",
            "FAIR\n",
            "Data\n",
            "Analysis\n",
            "Experiment\n",
            "Jupyter \n",
            "Notebook\n",
            "Acquire raw data\n",
            "Store data locally\n",
            "Tidy data\n",
            "Clean data\n",
            "Annotate data\n",
            "Deconvolve data\n",
            "Visualisations\n",
            "Outputs\n",
            "Markdown\n",
            "Code\n",
            "Share Data\n",
            "Name\n",
            "Repository set up\n",
            "Jupyter file\n",
            "Config file\n",
            "Readme file\n",
            "Make public\n",
            "GitHub\n",
            "Create Binder repository\n",
            "Link from GitHub\n",
            "Binder badge to GitHub \n",
            "Shareable Binder link\n",
            "Binder\n",
            "Private\n",
            "Open-access\n",
            "GitHub\n",
            "MetaboLights\n",
            "Metabolomics \n",
            "Workbench\n",
            "Private \n",
            "cloud\n",
            "Email\n",
            "Hard drive\n",
            "Account set up\n",
            "Share repository\n",
            "Fig. 3  Key elements required for FAIR data analysis, using Jupyter \n",
            "Notebooks and Binder deployment. A fishbone diagram describing \n",
            "the detailed requirements for FAIR data analysis in metabolomics. \n",
            "Experimental data are derived from typical metabolomics workflows \n",
            "and formatted appropriately for analysis. Data need to be shared, \n",
            "either privately (for pre-publication collaboration) or publicly (for \n",
            "open dissemination). The Jupyter Notebook contains all code, mark-\n",
            "down comments, outputs, and visualisations corresponding to the \n",
            "study. The Jupyter Notebook and other required files (such as Readme \n",
            "and configuration files) are compiled into a public GitHub repository. \n",
            "Finally, Binder is used to easily deploy and share the Jupyter Note-\n",
            "book\n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 9 of 16 \n",
            "125\n",
            "3.1.3  Binder\n",
            "Binder (mybin der.org) is an open source web service that \n",
            "allows users to deploy a GitHub repository comprising a \n",
            "collection of Jupyter Notebooks (with configuration files that \n",
            "describe the required computing environment) as a tempo-\n",
            "rary cloud-based virtual machine. The Binder deployment \n",
            "is accessible by web browser and includes the programming \n",
            "language and all necessary packages and data. As with all \n",
            "publicly-accessible cloud storage care must be taken if data \n",
            "are sensitive or private. Researchers can launch the virtual \n",
            "machine in their browser but, because the user environ-\n",
            "ment is temporary, once the session is closed all new results \n",
            "are lost. If changes are made, the user must download any \n",
            "changed files or output they wish to keep.\n",
            "3.2  Tutorials\n",
            "3.2.1  Tutorial 1: launching and using a Jupyter Notebook \n",
            "on Binder\n",
            "This tutorial demonstrates the use of computational note-\n",
            "books for transparent dissemination of data analysis work-\n",
            "flows and results. The tutorial steps though a metabolomics \n",
            "computational workflow implemented as a Jupyter Notebook \n",
            "and deployed on Binder. The workflow is designed to ana-\n",
            "lyse a deconvolved and annotated metabolomics data set \n",
            "(provided in an Excel workbook) and is an example of the \n",
            "standard data science axiom: Import, Tidy, Model, and \n",
            "Visualise.\n",
            "The Jupyter notebook for this tutorial is named Tutorial1.\n",
            "ipynb and is available at GitHub in the repository https ://\n",
            "githu b.com/cimcb /Metab Workfl owTu toria l. This repository \n",
            "can be downloaded (cloned) to the researcher’s own com-\n",
            "puter, or run on the Binder service. In the text we assume \n",
            "that the tutorial is being run using the Binder service. To \n",
            "open the notebook on Binder, go to the tutorial homepage: \n",
            "https ://cimcb .githu b.io/Metab Workfl owTu toria l and click \n",
            "on the topmost “Launch Binder” icon to “launch the tuto-\n",
            "rial environment in the cloud”. It will take a short while for \n",
            "Binder to build and deploy a new temporary virtual machine. \n",
            "Once this is ready the Jupyter notebook landing page will \n",
            "show the files present in this copy of the GitHub repository \n",
            "(Supplementary Fig. 1).\n",
            "The tutorial workflow analysis interrogates a published \n",
            "dataset used to discriminate between samples from gastric \n",
            "cancer and healthy patients (Chan et al. 2016). The dataset is \n",
            "available in the Metabolomics Workbench database (http://\n",
            "www.metab olomi cswor kbenc h.org, Project ID PR000699). \n",
            "For this tutorial, the data are stored in the Excel workbook \n",
            "Fig. 4  Example Jupyter Notebook Screenshot. At the top of the page, \n",
            "there is the Jupyter menu bar and ribbon of action buttons. The main \n",
            "body of the notebook then displays text and code cells, and any out-\n",
            "puts from code execution. This screenshot taken near the end of Tuto-\n",
            "rial 1 when the partial least squares discriminant analysis model is \n",
            "being evaluated. Three plots are generated, showing comparisons of \n",
            "the performance of the model on training and holdout test datasets: \n",
            "a violin plot showing the distribution of known positive and nega-\n",
            "tive in both training and test sets, and the class cut-off (dotted line); \n",
            "probability density functions for positive and negative classes in the \n",
            "training and test sets (the training set datapoints are rendered as more \n",
            "opaque); ROC curves of model performance on training (with 95% \n",
            "CI) and test set\n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 10 of 16\n",
            "GastricCancer_NMR.xlsx using the Tidy Data framework \n",
            "(Wickham 2014): each variable is a column, each observa-\n",
            "tion is a row, and each type of observational unit is a table. \n",
            "The data are split into two linked tables. The first, named \n",
            "‘Data’, contains data values related to each observation. i.e. \n",
            "metabolite concentrations  M1 …  Mn, together with meta-\n",
            "data such as: ‘sample type’, ‘sample identifier’ and ‘outcome \n",
            "class’. The second, named ‘Peak’, contains data that links \n",
            "each metabolite identifier  (Mi) to a specific annotation and \n",
            "optional metadata (e.g. mass, retention time, MSI identifica-\n",
            "tion level, number of missing values, quality control meas-\n",
            "ures, etc.). The Excel file can also be downloaded from the \n",
            "Binder virtual machine for inspection on your own machine \n",
            "by selecting the checkbox next to the filename and clicking \n",
            "on the Download button in the top menu (Supplementary \n",
            "Fig. 1).\n",
            "To begin the tutorial, click on the Tutorial1.ipynb file-\n",
            "name (Supplementary Fig. 1). This will open a new tab in \n",
            "your browser presenting the Jupyter notebook (Supplemen-\n",
            "tary Fig. 2). At the top of the page there is a menu bar and \n",
            "ribbon of action buttons similar to those found in other GUI-\n",
            "based software, such as Microsoft Word. The interface is \n",
            "powerful, and it is worth taking time to become familiar with \n",
            "it, but for this tutorial only the “Run” button and the “Cell” \n",
            "and “Kernel” drop down menus are required.\n",
            "The rest of the page is divided into “code cells” and “text \n",
            "cells”. The “text cells” briefly outline the context and com-\n",
            "putation of the “code cells” beneath them. Code and text \n",
            "cells can be distinguished by their background colour (code \n",
            "cells are slightly grey, text cells are slightly red), by the text \n",
            "formatting (code cells have a fixed-width font, text cells have \n",
            "word processor-like formatting), and the “In []:” marker text \n",
            "is present next to each code cell.\n",
            "To run a single code cell, first select it by clicking any-\n",
            "where within the cell, which will then be outlined by a green \n",
            "box (if you select a text cell, this box is blue—Supplemen-\n",
            "tary Fig. 3). Once a cell is selected, the code in the cell \n",
            "can be executed by clicking on the “Run” button in the top \n",
            "menu. Multiple cells can also be run in sequence by choos-\n",
            "ing options from the dropdown list in the “Cell” menu item. \n",
            "The options include “Run All” (runs all the cells in the note-\n",
            "book, from top to bottom), and “Run all below” (run all cells \n",
            "below the current selection). These can be used after chang-\n",
            "ing the code or values in one cell to recalculate the contents \n",
            "of subsequent cells in the notebook.\n",
            "The “computational engine” that executes the code con-\n",
            "tained in a notebook document is called the kernel, and it \n",
            "runs continually in the background while that notebook \n",
            "is active. When you run a code cell, that code is executed \n",
            "by the kernel and any output is returned back to the note-\n",
            "book to be displayed beneath the cell. The kernel stores \n",
            "the contents of variables, updating them as each cell is \n",
            "run. It is always possible to return to a “clean” state by \n",
            "choosing one of the “Restart Kernel” options from the \n",
            "“Kernel” menu item’s dropdown list. Selecting “Restart & \n",
            "Run All” from the “Kernel” dropdown menu will restart \n",
            "the kernel and run all cells in order from the start to the \n",
            "end of the notebook.\n",
            "Beginning from a freshly-loaded Tutorial1.ipynb note-\n",
            "book in the Binder, clicking on “Cell->Run All” or “Kernel-\n",
            ">Restart & Run All” will produce a fully executed notebook \n",
            "that matches the output in the static supplementary html file \n",
            "Tutorial1.html (cimcb .githu b.io/Metab Workfl owTu toria l/\n",
            "Tutor ial1.html). Choosing “Restart and Clear Outputs” from \n",
            "the “Kernel” dropdown menu, will reset the notebook and \n",
            "clear all data from memory and remove any outputs, restor-\n",
            "ing its original state.\n",
            "The tutorial can be completed by reading the text cells \n",
            "in the notebook and inspecting, then running, the code in \n",
            "the corresponding code cells. This is an example of “Lit-\n",
            "erate Programming” that weaves traditional computing \n",
            "source code together with a human-readable, natural lan-\n",
            "guage description of the program logic (Knuth 1984). The \n",
            "notebook interface makes notable advances on the original \n",
            "proposition for literate programming that are used in this \n",
            "tutorial, the most significant of which is that the output of \n",
            "running the code is also incorporated into the document. The \n",
            "browser interface allows for further enhancements, such as \n",
            "hyperlinks to external webpages for explanations and fur-\n",
            "ther reading about technical terms, embedded interactive \n",
            "spreadsheet-like representation of large datasets (e.g. sec-\n",
            "tion 2. Load Data and Peak Sheet), and embedded interactive \n",
            "graphical output (e.g. section 4. PCA Quality Assessment).\n",
            "3.2.2  Tutorial 2: interacting with and editing a Jupyter \n",
            "Notebook on Binder\n",
            "The second tutorial is interactive and showcases the utility \n",
            "of computational notebooks for both open collaboration and \n",
            "experiential education in metabolomics data science. Tuto-\n",
            "rial 2 is accessed on GitHub through the same process as \n",
            "described for Tutorial 1. To open the notebook on Binder, \n",
            "go to the tutorial homepage: https ://cimcb .githu b.io/Metab \n",
            "Workfl owTu toria l and click on the topmost “Launch Binder” \n",
            "icon to “launch the tutorial environment in the cloud”, then \n",
            "click the Tutorial2.ipynb link on the Jupyter landing page. \n",
            "This will present a new tab in your browser containing the \n",
            "second tutorial notebook. The functionality of this notebook \n",
            "is identical to Tutorial 1, but now the text cells have been \n",
            "expanded into a comprehensive interactive tutorial. Text \n",
            "cells, with a yellow background, provide the metabolomics \n",
            "context and describe the purpose of the code in the follow-\n",
            "ing code cell. Additional coloured text boxes are placed \n",
            "throughout the workflow to help novice users navigate and \n",
            "understand the interactive principles of a Jupyter Notebook:\n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 11 of 16 \n",
            "125\n",
            "3.2.2.1 Action (red background labelled with  ‘gears’ \n",
            "icon) Red boxes provide suggestions for changing the \n",
            "behaviour of the subsequent code cell by editing (or sub-\n",
            "stituting) a line of code. For example, the first red cell \n",
            "describes how to change the input dataset by changing the \n",
            "path to the source Excel file.\n",
            "3.2.2.2 Interaction \n",
            "(green \n",
            "background \n",
            "with  ‘mouse’ \n",
            "icon) Green boxes provide suggestions for interacting with \n",
            "the visual results generated by a code cell. For example, the \n",
            "first green box in the notebook describes how to sort and \n",
            "colour data in the embedded data tables.\n",
            "3.2.2.3 Notes (blue background with ‘lightbulb’ icon) Blue \n",
            "boxes provide further information about the theoretical rea-\n",
            "soning behind the block of code or a given visualisation. \n",
            "This information is not essential to understand Jupyter Note-\n",
            "books but may be of general educational utility and interest \n",
            "to new metabolomics data scientists.\n",
            "To complete the tutorial, first execute the notebook by \n",
            "selecting the “Restart & Run All” option in the “Kernel” \n",
            "dropdown menu. Move through the notebook one cell at a \n",
            "time reading the text and executing the code cells. When \n",
            "prompted, complete one (or multiple) modifications sug-\n",
            "gested in each ‘action’ box, and the click “Run all below” \n",
            "from the “Cell” dropdown menu, observing the changes in \n",
            "cell output for all the subsequent cells. Further guidance is \n",
            "included in the notebook itself.\n",
            "It is possible to save the edited notebook to the Binder \n",
            "environment, but any changes made to the notebook during \n",
            "the tutorial are lost when the Binder session ends. To keep \n",
            "changes made to the tutorial notebook or its output, modi-\n",
            "fied files must be downloaded to your local computer before \n",
            "you end the session. Modified files can also be downloaded \n",
            "from the Jupyter landing page. To download files, click the \n",
            "checkbox next to each file you wish to download, and then \n",
            "click the ‘Download’ button from the top menu.\n",
            "3.2.3  Tutorial 3: downloading and installing a Jupyter \n",
            "Notebook on a local machine\n",
            "Jupyter Notebooks can be run on a standard laptop or desk-\n",
            "top computer in a number of different ways, depending on \n",
            "the operating system. The Anaconda distribution provides a \n",
            "unified, platform-independent framework for running note-\n",
            "books and managing Conda virtual environments that is con-\n",
            "sistent across multiple operating systems, so for convenience \n",
            "we will use the Anaconda interface in these tutorials.\n",
            "To install the Anaconda distribution, first download \n",
            "the Python 3.x Graphical Installer package from the Ana-\n",
            "conda webpage (https ://www.anaco nda.com/distr ibuti on/) \n",
            "then open the installer and follow the instructions to com-\n",
            "pete the installation (https ://docs.anaco nda.com/anaco nda/\n",
            "insta ll/). Be sure to download the installer package specific \n",
            "to your computer’s operating system (e.g. macOS, Micro-\n",
            "soft Windows or Linux). When the process is completed, \n",
            "the “Anaconda Navigator” application will be installed in \n",
            "your applications folder.\n",
            "To start Jupyter on your machine first launch the Ana-\n",
            "conda Navigator application. This will display a home \n",
            "screen with a sidebar menu on the left-hand side and the \n",
            "main area showing a panel of application icons, with short \n",
            "descriptions. Locate the Jupyter Notebook application and \n",
            "icon in this panel and click the “launch” button under the \n",
            "icon. This will start a Jupyter web server and open the \n",
            "Jupyter landing page in your default web browser. To run \n",
            "an existing Jupyter notebook, navigate to the appropriate \n",
            "folder on your computer’s filesystem in the Jupyter landing \n",
            "page, and click on the notebook (.ipynb) file you wish to \n",
            "open. To end a Jupyter session, click on the “quit” button \n",
            "in the top right-hand corner of the Jupyter landing page. \n",
            "Quit now if you have been working along.\n",
            "To run the Tutorial notebooks, we need to download \n",
            "the tutorial repository containing those notebooks from \n",
            "GitHub and set up a local “virtual environment” that con-\n",
            "tains the programming libraries and software tools neces-\n",
            "sary to run the code cells in the notebooks.\n",
            "To download the notebook and associated files from the \n",
            "Github repository page (https ://githu b.com/cimcb /Metab \n",
            "Workfl owTu toria l), click on the green button labelled \n",
            "“clone or download” and choose the option to “Download \n",
            "ZIP”. Save the zip file (MetabWorkflowTutorial-master.\n",
            "zip) in a convenient location. Extract the zip file to create \n",
            "a new folder in the same location as the .zip file, called \n",
            "“MetabWorkflowTutorial-master”. The contents of this \n",
            "folder are the files visible in the repository at the GitHub \n",
            "site. We will refer to this folder as the “repository root”, \n",
            "or just “root”.\n",
            "The Jupyter Notebooks in the repository require several \n",
            "Python packages to be installed in order to be run suc-\n",
            "cessfully. It would be possible to install these on the local \n",
            "computer so that they are visible to, and accessible by, all \n",
            "notebooks on the computer. However, it is often the case \n",
            "that different repositories and projects require alternative, \n",
            "incompatible versions of these packages. So, in practice, it \n",
            "is not usually possible to install a single set of packages that \n",
            "meets the needs of all the projects that a user would want to \n",
            "run. A technical solution to this is to create a new “virtual \n",
            "environment” that contains only the packages necessary for a \n",
            "project to run, and keeps them separate (“sandboxes” them) \n",
            "from any other projects. Environments can be created when \n",
            "required, and deleted when no longer necessary, without \n",
            "affecting other projects or the operation of the computer. \n",
            "It is good practice to create a new virtual environment for \n",
            "each project, and typical that multiple such environments are \n",
            "set up, and exist simultaneously on the same computer. The \n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 12 of 16\n",
            "Anaconda Navigator application provides an interface for \n",
            "creating and managing these virtual environments.\n",
            "To create a new virtual environment for the tutorial, \n",
            "first open the Anaconda Navigator application and click on \n",
            "“Environments” in the left-hand sidebar. The main panel \n",
            "will change to list any virtual environments that have been \n",
            "created using Anaconda. If no environments have been cre-\n",
            "ated only “base (root)” will be listed. To the right of each \n",
            "virtual environment Anaconda Navigator lists the packages \n",
            "that have been installed in that environment.\n",
            "It is common to create a new environment “from scratch” \n",
            "by specifying individual packages in the Anaconda Navi-\n",
            "gator, but for this tutorial we will use a configuration file \n",
            "called “environment.yml” that is part of the GitHub reposi-\n",
            "tory. This file describes all the packages that are necessary \n",
            "to reproduce an environment for running the tutorial note-\n",
            "books. To create a new environment from this configura-\n",
            "tion file, click on “Import” (at the bottom of the main panel \n",
            "of Anaconda Navigator) and navigate to the repository root \n",
            "folder. By default Anaconda Navigator expects configuration \n",
            "files with “.yaml” or “.yml” file extensions, so only the file \n",
            "named “environment.yml” should be highlighted in the file \n",
            "dialog box. Select this file and click “Open”. The “Import \n",
            "new environment” dialogue box will have autocompleted \n",
            "the “Name:” field for the new environment (“MetabWork-\n",
            "flowTutorial”). To complete creation of the new environ-\n",
            "ment, click on the “Import” button. Anaconda Navigator \n",
            "will show a progress bar in the main panel as it creates the \n",
            "new environment.\n",
            "Once the environment has been created, click on the \n",
            "“Home” icon in the left-hand sidebar. In the main panel, \n",
            "the dropdown should now read “Applications on [Metab-\n",
            "WorkflowTutorial]”, which indicates that the MetabWork-\n",
            "flowTutorial environment which was just created is now \n",
            "active. If “MetabWorkflowTutorial” is not visible, click on \n",
            "the dropdown menu and select that environment. Click on \n",
            "the “Launch” button under Jupyter Notebook in the main \n",
            "panel, to launch Jupyter in your web browser.\n",
            "The Jupyter landing page will start in your home folder. \n",
            "To use the tutorial notebooks, navigate to the repository root. \n",
            "The notebooks for Tutorial 1 and 2 can now be run on your \n",
            "own computer, just as on Binder, by selecting the appro-\n",
            "priate notebook file. However any output or changes to the \n",
            "contents of a notebook file will now be saved persistently in \n",
            "the local computer and can be reused at any time.\n",
            "As an alternative you may wish to try to create a vir-\n",
            "tual environment and launch Jupyter in your web browser \n",
            "through a terminal window (command window). To do this \n",
            "open the terminal window (type ‘terminal’ in your comput-\n",
            "er’s search box), then type the following five lines of code:\n",
            "git clone https://github.com/cimcb/\n",
            "MetabWorkflowTutorial\n",
            "cd MetabWorkflowTutorial\n",
            "conda env create -f environment.yml\n",
            "conda activate MetabWorkflowTutorial\n",
            "jupyter notebook\n",
            "Line one creates an exact copy of the github file directory \n",
            "on your local machine in the folder ‘MetabWorkflowTuto-\n",
            "rial’. Line two moves you into that folder. Line three creates \n",
            "the virtual environment called “MetabWorkflowTutorial” \n",
            "using the contents of the environment.yml file. Line four \n",
            "activates the virtual environment. Line five launches a local \n",
            "Jupyter notebook server and opens the Jupyter landing page \n",
            "in your web browser, from which you can run the tutorials.\n",
            "To close the local Jupyter notebook server press \n",
            "“<control>c” twice in the terminal window and it will ask \n",
            "you to confirm the action. You may then close the virtual \n",
            "environment by typing:\n",
            "conda deactivate\n",
            "When you no longer need the virtual environment, the \n",
            "following will delete it from your computer:\n",
            "conda remove –name MetabWorkflowTuto-\n",
            "rial –all\n",
            "If you created a virtual environment using Anaconda \n",
            "Navigator you will have to delete the environment before \n",
            "creating a fresh version.\n",
            "3.2.4  Tutorial 4: creating a new Jupyter Notebook \n",
            "on a local computer\n",
            "Tutorial 4 builds on tutorial 3. Please ensure that the Ana-\n",
            "conda Python distribution is installed on your computer.\n",
            "In this tutorial we will create a new Jupyter notebook \n",
            "that demonstrates the use of visualisation methods available \n",
            "in Anaconda Python without the need to install additional \n",
            "third-party packages. We will upload a generic metabo-\n",
            "lomics data set and write code to produce four graphical \n",
            "outputs:\n",
            "(1) A histogram of the distribution of  QCRSD across the \n",
            "data set.\n",
            "(2) A kernel density plot of  QCRSD vs. D-ratio across the \n",
            "data set.\n",
            "(3) A PCA scores plot of the data set labelled by sample \n",
            "type.\n",
            "(4) A bubble scatter plot of molecular mass vs. retention \n",
            "time, with bubble size proportional to  QCRSD\n",
            "The data set included in this tutorial is previously unpub-\n",
            "lished, and of arbitrary biological value. It describes serum \n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 13 of 16 \n",
            "125\n",
            "data acquired using a C18+ LC–MS platform consisting of \n",
            "3084 unidentified peaks and 91 samples. Of the 91 sam-\n",
            "ples, 23 are pooled QCs injected every 5th sample across \n",
            "the experimental run. The Peak table contains informa-\n",
            "tion on the molecular mass, retention time of each detected \n",
            "metabolite, and the associated QCRSD and D-ratio values \n",
            "calculated following recommended quality control proce-\n",
            "dures (Broadhurst et al. 2018). The data are presented in an \n",
            "Excel file using the previously-described “tidy data” format.\n",
            "Tutorial 4 is available in a GitHub repository at https ://\n",
            "githu b.com/cimcb /Metab Simpl eQcVi z. Download and unzip \n",
            "the repository to a folder on your own computer, using the \n",
            "method described in Tutorial 3 (the location of this folder \n",
            "will now be the “repository root”). This copy (clone) of the \n",
            "repository is for reference only as we will be recreating the \n",
            "contents of this directory under a different name as we move \n",
            "through this tutorial and Tutorial 5.\n",
            "First create a new Jupyter notebook. To do this, start the \n",
            "Anaconda Navigator application if it is not already open. \n",
            "Ensure that “[base (root)]” is selected in the “Applications \n",
            "on” dropdown list of the main panel, then launch Jupyter \n",
            "Notebook. This will start a new Jupyter notebook server in \n",
            "your browser and show files from the home directory on the \n",
            "landing page. Navigate to the repository root (the “Metab-\n",
            "SimpleQcViz” folder). To create a new notebook, click on \n",
            "the “New” button in the top right corner of the page. This \n",
            "will list supported Jupyter Notebook languages in the drop-\n",
            "down. Select “Python 3” from this list. A new tab will open \n",
            "in your browser, showing a blank notebook called “Untitled” \n",
            "(at the top of the page). Rename the notebook by clicking on \n",
            "the text “Untitled” and replacing it with “myExample”. This \n",
            "will create a new file in the repository called “myExample.\n",
            "ipynb”\n",
            "When the “myExample.ipynb” notebook is launched, it \n",
            "contains a single empty code cell. We will use this cell to \n",
            "add a title to the notebook. To do this we need to convert the \n",
            "cell type to be a Markdown cell, then type a header in the \n",
            "cell, and execute it. First, select the empty cell by clicking \n",
            "anywhere within the cell. To convert the cell type, click on \n",
            "the dropdown field marked “Code” in the top menu bar and \n",
            "select “Markdown”. The “In[]:” prompt should disappear \n",
            "from the left-hand side of the cell. Now click inside the cell \n",
            "to see the flashing cursor that indicates the cell is ready to \n",
            "accept input. Type “# Tutorial 4” and click on the “Run” but-\n",
            "ton in the top menu. The formatting of the first cell should \n",
            "change, and a new code cell should appear beneath it.\n",
            "In the new code cell, we will place Python code that:\n",
            "1. Imports the Pandas package (necessary to load the Excel \n",
            "spreadsheet).\n",
            "2. Loads the dataset into variables called “data” and \n",
            "“peak”.\n",
            "3. Report the number of rows and column in the tables.\n",
            "4. Displays the first few lines of the resulting table.\n",
            "The required code is provided in the static supplementary \n",
            "html file Tutorial4.html (https ://cimcb .githu b.io/Metab Simpl \n",
            "eQcVi z/Tutor ial4.html) and “Tutorial4.ipynb” notebook and \n",
            "can be copy-and-pasted or typed in manually, as preferred. \n",
            "When the code is complete, click on the “Run” button again \n",
            "to execute the cell. On completion, two tables should be \n",
            "visible below the code cell (one for “data”, one for “peak”), \n",
            "and a new empty code cell should be placed beneath this.\n",
            "Next we add the code required to draw a histogram of the \n",
            "RSD values across all the detected peaks in this data set. \n",
            "Using the Tutorial4.html file as a guide, add in the required \n",
            "explanatory text and Python code and click on the “Run” \n",
            "button after each step.\n",
            "Continue adding in the remaining explanatory text and \n",
            "Python code using the Tutorial4.html file. After completion \n",
            "you will have a Jupyter notebook that takes a metabolomics \n",
            "dataset through the process of generating diagnostic plots \n",
            "for quality control. Once you are satisfied with the state of \n",
            "the notebook, it can be saved by clicking on the floppy disk \n",
            "icon (far left on the menu). The notebook can then be closed \n",
            "by clicking “File” and then “Close and Halt” from the top \n",
            "Jupyter menu. The notebook tab will be closed, showing the \n",
            "Jupyter landing page. The Jupyter session can be closed by \n",
            "clicking on “Quit” on the Jupyter landing page tab of your \n",
            "web browser (this tab may not close automatically).\n",
            "3.2.5  Tutorial 5: deploying a Jupyter Notebook on Binder \n",
            "via GitHub\n",
            "Tutorial 5 builds on tutorial 3 and 4. To complete this tuto-\n",
            "rial, we will create a new GitHub repository. A GitHub \n",
            "account is required for this. If you do not already have a \n",
            "GitHub account, please follow the instructions on GitHub \n",
            "at https ://help.githu b.com/en/artic les/signi ng-up-for-a-new-\n",
            "githu b-accou nt.\n",
            "To create a new repository, log into the GitHub site (if \n",
            "you are not already logged in) and navigate to your profile \n",
            "page (https ://githu b.com/<youru serna me>), then click on \n",
            "the “Repositories” link at the top of the page. To start a \n",
            "new repository, click on the “New” button at the top right \n",
            "of the page. This will open a new page titled “Create a new \n",
            "repository.” Each repository requires a name, and this should \n",
            "be entered into the “Repository name” field; use the name \n",
            "“JupyterExample”. Beneath the Repository Name field \n",
            "there is an optional Description box, and then below this a \n",
            "choice of public or private repository. Ensure that the ‘Pub-\n",
            "lic’ option is chosen. Select the checkbox to “Initialize this \n",
            "repository with a README” (this is a file in which you \n",
            "will write useful information about the repository, later). \n",
            "Below this is the option to “Add a license” file. There are \n",
            "many alternative licences to choose from (https ://choos ealic \n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 14 of 16\n",
            "ense.com/), and the choice for your own projects may be \n",
            "constrained by funder, home organisation, or other legal con-\n",
            "siderations. We strongly recommend that all projects carry \n",
            "a suitable licence, and that you add the MIT License to this \n",
            "tutorial repository. Now, to create the repository, click the \n",
            "“Create repository” button.\n",
            "On successful creation of the repository, GitHub will pre-\n",
            "sent the new repository’s home page (this will be at https ://\n",
            "githu b.com/<youru serna me>/Jupyt erExa mple), with some \n",
            "options for “Quick setup”. Under the “Quick setup” notice, \n",
            "the LICENSE and README.md file will be shown, and \n",
            "clicking on either will open them. The README.md file for \n",
            "a repository is automatically displayed on the homepage, but \n",
            "in this case, it is empty (we can add text later).\n",
            "Now we need to add the new Jupyter notebook and the \n",
            "Excel data file from tutorial 4 to the repository. We will \n",
            "do this using the GitHub “Upload files” interface, though \n",
            "there are several other ways to perform this action. To \n",
            "use the GitHub interface, click on the ‘Upload files’ but-\n",
            "ton and either drag files from your computer, or click on \n",
            "“choose your files” to select files with a file dialogue box. \n",
            "Add the ‘myExample.ipynb’ and ‘data.xlsx’ files from your \n",
            "repository root. These files will be placed in the “staging \n",
            "area”, visible on the webpage but not yet committed to the \n",
            "repository.\n",
            "GitHub imposes version control as a form of best prac-\n",
            "tice on the repositories it hosts. One of the features of ver-\n",
            "sion control best practice is that a description of the changes \n",
            "made to a repository should accompany every “commit” to \n",
            "that repository. To do this, enter the text “Add data and note-\n",
            "book via upload” to the top field under “Commit changes.” \n",
            "Then, to commit the files to the repository, click on the \n",
            "“Commit changes” button.\n",
            "Now that there is a publicly hosted GitHub repository \n",
            "containing a notebook and dataset, we are nearly ready to \n",
            "make the notebook available interactively through Binder. \n",
            "The final necessary component required is a configuration \n",
            "file. This file is vital, as it defines the environment Binder \n",
            "will build, with a specified programming language and all \n",
            "the necessary packages for the notebook to successfully \n",
            "operate. This configuration file is an Anaconda YAML file \n",
            "called ‘environment.yml’ and it contains a list of dependen-\n",
            "cies (the programming language version and a list of pack-\n",
            "ages used in the notebook) and channels (the location of \n",
            "these resources in the Anaconda cloud library). Detailed \n",
            "consideration of how to create these files is beyond the scope \n",
            "of the tutorial. Upload the environment.yml file from Tuto-\n",
            "rial 4 (it is also included in the Supplementary File, to cut \n",
            "and paste if required) to the repository in the same way that \n",
            "the notebook and data files were uploaded.\n",
            "We are now ready to build and launch a Binder virtual \n",
            "machine for this repository. To do this, open https ://mybin \n",
            "der.org in a modern web browser. The landing page presents \n",
            "a set of fields to be completed for Binder to build a virtual \n",
            "machine. The minimal requirement is to specify a GitHub \n",
            "repository URL in the “GitHub repository name or URL” \n",
            "field Enter the path to the home page of your repository \n",
            "(https ://githu b.com/<youru serna me>/Jupyt erExa mple) in \n",
            "this field, and click on the ‘Launch’ button. Binder will use \n",
            "the configuration file in the root directory to build and store \n",
            "a Docker image for your repository. This process often takes \n",
            "several minutes.\n",
            "Once the Binder repository is built, the URL shown in \n",
            "the field “Copy the URL below and share your Binder with \n",
            "others” (here: https ://mybin der.org/v2/gh/<youru serna me>/\n",
            "Jupyt erExa mple/maste r) can be shared with colleagues. \n",
            "A button to launch the Binder can also be added into the \n",
            "README file on GitHub (we also strongly recommend \n",
            "this). Anyone using this URL in their browser, will be pro-\n",
            "vided with an individual interactive session (1 CPU, 2 GB \n",
            "RAM running on Google Cloud) making available the note-\n",
            "books of your repository in an interactive and editable form.\n",
            "Congratulations, you have created your first Binder note-\n",
            "book! Now share it with your colleagues!\n",
            "It is important to remind users that data uploaded to a \n",
            "public GitHub repository is indeed public. If the user wants \n",
            "to share Jupyter Notebooks but not any associated metabo-\n",
            "lomics data (or other sensitive data) then clear instructions \n",
            "on how to securely access and download the data needs to be \n",
            "included in the notebook text, and the location of that down-\n",
            "loaded data be included in the requisite notebook code block \n",
            "(this could be a local hard drive, or uploaded to Binder while \n",
            "in session). If institutional security concerns preclude using \n",
            "a collaborative workspace such as Binder, then alternative \n",
            "cloud solutions such as Microsoft Azure can be investigated. \n",
            "Before doing so it is probably best that to consult with your \n",
            "institute IT representative.\n",
            "4  Summary\n",
            "Due to the rate at which data are generated and new analy-\n",
            "sis and visualisation methods are developed, the omics \n",
            "sciences have become highly vulnerable to irreproducibil-\n",
            "ity. In attempt to ameliorate this, the metabolomics com-\n",
            "munity has made several efforts to align with FAIR data \n",
            "standards in the areas of open data formats, data reposi-\n",
            "tories, online spectral libraries, and metabolite databases. \n",
            "While there are also a number of open options for data \n",
            "analysis, these tend to exist as prescriptive and inflexible \n",
            "workflows that inadvertently enable users to apply data \n",
            "science methods without fully understanding their under-\n",
            "lying principles and assumptions. For FAIR data science to \n",
            "exist in metabolomics, presentation of methods and results \n",
            "needs to be rapid, transparent, reusable, and recoverably \n",
            "Toward collaborative open data science in metabolomics using Jupyter Notebooks and cloud…\n",
            "1 3\n",
            "Page 15 of 16 \n",
            "125\n",
            "attached to published work. Furthermore, any framework \n",
            "enabling this must be intuitive and accessible to compu-\n",
            "tational novices.\n",
            "In this tutorial review, we have illustrated one pos-\n",
            "sible solution for achieving open, transparent, yet intui-\n",
            "tive data science within the metabolomics community. \n",
            "Jupyter Notebooks are an open-source, interactive web \n",
            "tool for creating seamless integration of text, code, and \n",
            "outputs (tables, figures) into a single live executable \n",
            "document. When used alongside data repositories, such \n",
            "as GitHub, and open cloud-based deployment services, \n",
            "such as Binder, these computational notebooks can greatly \n",
            "enhance transparent dissemination of data science meth-\n",
            "ods and results during the publication process. In addition \n",
            "to the benefit of increased transparency, computational \n",
            "notebooks provide a valuable tool for open collaboration. \n",
            "Rather than exchanging multiple individual data, code, \n",
            "methods, and results files, computational notebook envi-\n",
            "ronments provide a single mechanism for collaborators \n",
            "(both within and beyond a single research group) to share \n",
            "and interact with the data science workflow. Moreover, this \n",
            "interactive nature, combined with the ability to provide \n",
            "extensive documentation, provides a valuable opportunity \n",
            "for enhanced learning in the computer programming and \n",
            "data science contexts. Given that they are increasingly rec-\n",
            "ognised as being foundational to contemporary research, \n",
            "it is imperative that scientists continue to enhance these \n",
            "skills over the duration their career. This open and interac-\n",
            "tive framework enables scientists to continue to learn and \n",
            "also keep up-to-date with latest data science methods and \n",
            "trends without reinstalling the wheel.\n",
            "Acknowledgements This work was partly funded through an Austral-\n",
            "ian Research Council funded LIEF Grant No. (LE170100021).\n",
            "Author contributions All authors contributed equally to this work. \n",
            "KMM developed the code for the Python cimcb-lite package. KMM \n",
            "& DIB developed the tutorial Jupyter notebook code. LP and SNR \n",
            "edited the learning tutorials to align to current pedagogical best prac-\n",
            "tices. KMM wrote the initial draft of the review section. All authors \n",
            "edited the manuscript.\n",
            "Data and software availability The metabolomics and metadata used \n",
            "in this paper were retrieved from Metabolights (https ://www.ebi.\n",
            "ac.uk/metab oligh ts/) study identifier: MTBLS290, and Metabolomics \n",
            "Workbench (https ://www.metab olomi cswor kbenc h.org/) project id: \n",
            "PR000699. This data were converted from the original data format to \n",
            "a clean format compliant with the Tidy Data framework, this is avail-\n",
            "able at the CIMCB GitHub project page (https ://githu b.com/CIMCB ).\n",
            "All software developed for this paper is available at the CIMCB \n",
            "GitHub project page (https ://githu b.com/CIMCB ).\n",
            "Compliance with ethical standards \n",
            "Conflict of interest The authors have no disclosures of potential con-\n",
            "flicts of interest related to the presented work.\n",
            "Research involving human and animal participants No research involv-\n",
            "ing human or animal participants was performed in the construction \n",
            "of this manuscript.\n",
            "Open Access This article is distributed under the terms of the Crea-\n",
            "tive Commons Attribution 4.0 International License (http://creat iveco \n",
            "mmons .org/licen ses/by/4.0/), which permits unrestricted use, distribu-\n",
            "tion, and reproduction in any medium, provided you give appropriate \n",
            "credit to the original author(s) and the source, provide a link to the \n",
            "Creative Commons license, and indicate if changes were made.\n",
            "References\n",
            "Afgan, E., Baker, D., Batut, B., van den Beek, M., Bouvier, D., Cech, \n",
            "M., et al. (2018). The Galaxy platform for accessible, reproducible \n",
            "and collaborative biomedical analyses: 2018 update. Nucleic Acids \n",
            "Research, 46, W537–W544.\n",
            "Anderson, L. W., Krathwohl, D. R., Airasian, P. W., Cruikshank, K. \n",
            "A., Mayer, R. E., Pintrich, P. R., et al. (2001). A taxonomy for \n",
            "learning, teaching, and assessing: A revision of Bloom’s taxon-\n",
            "omy of educational objectives (abridged ed.). White Plains, NY: \n",
            "Longman.\n",
            "Baker, M. (2016). 1,500 scientists lift the lid on reproducibility. Nature, \n",
            "533, 452–454.\n",
            "Baumer, B., Cetinkaya-Rundel, M., Bray, A., Loi, L. and Horton, N.J. \n",
            "(2014) R markdown: Integrating a reproducible analysis tool into \n",
            "introductory statistics, Technology Innovations in Statistics Edu-\n",
            "cation, 8\n",
            "Beaker X Development Team (2018). Beaker X. Retrieved May 1, \n",
            "2019, from http://beake rx.com/.\n",
            "Bokeh Development Team (2018). Bokeh: Python library for interac-\n",
            "tive visualization. Retrieved May 1, 2019, from http://www.bokeh \n",
            ".pydat a.org.\n",
            "Broadhurst, D., Goodacre, R., Reinke, S. N., Kuligowski, J., Wilson, I. \n",
            "D., Lewis, M. R., et al. (2018). Guidelines and considerations for \n",
            "the use of system suitability and quality control samples in mass \n",
            "spectrometry assays applied in untargeted clinical metabolomic \n",
            "studies. Metabolomics, 14, 72.\n",
            "Broadhurst, D. I., & Kell, D. B. (2006). Statistical strategies for avoid-\n",
            "ing false discoveries in metabolomics and related experiments. \n",
            "Metabolomics, 2, 171–196.\n",
            "Chan, A. W., Mercier, P., Schiller, D., Bailey, R., Robbins, S., Eurich, \n",
            "D. T., et al. (2016). (1)H-NMR urinary metabolomic profiling for \n",
            "diagnosis of gastric cancer. British Journal of Cancer, 114, 59–62.\n",
            "Considine, E. C., Thomas, G., Boulesteix, A. L., Khashan, A. S., & \n",
            "Kenny, L. C. (2017). Critical review of reporting of the data analy-\n",
            "sis step in metabolomics. Metabolomics, 14, 7.\n",
            "Davidson, R. L., Weber, R. J. M., Liu, H., Sharma-Oates, A., & Viant, \n",
            "M. R. (2016). Galaxy-M: A Galaxy workflow for processing and \n",
            "analyzing direct infusion and liquid chromatography mass spec-\n",
            "trometry-based metabolomics data. GigaScience, 5, 10.\n",
            "Gehlenborg, N., O’Donoghue, S. I., Baliga, N. S., Goesmann, A., \n",
            "Hibbs, M. A., Kitano, H., et al. (2010). Visualization of omics \n",
            "data for systems biology. Nature Methods, 7, S56–S68.\n",
            "Giacomoni, F., Le Corguillé, G., Monsoor, M., Landi, M., Pericard, P., \n",
            "Pétéra, M., et al. (2015). Workflow4Metabolomics: A collabora-\n",
            "tive research infrastructure for computational metabolomics. Bio-\n",
            "informatics (Oxford, England), 31, 1493–1495.\n",
            "GitHub (2019). About GitHub. Retrieved April 30, 2019, from https \n",
            "://githu b.com/about .\n",
            "Goodacre, R., Broadhurst, D., Smilde, A. K., Kristal, B. S., Baker, \n",
            "J. D., Beger, R., et al. (2007). Proposed minimum reporting \n",
            " \n",
            "K. M. Mendez et al.\n",
            "1 3\n",
            "125 \n",
            "Page 16 of 16\n",
            "standards for data analysis in metabolomics. Metabolomics, 3, \n",
            "231–241.\n",
            "Gruber, J. (2004). Markdown. Retrieved April 30, 2019, from https ://\n",
            "darin gfire ball.net/proje cts/markd own/.\n",
            "Haug, K., Salek, R. M., Conesa, P., Hastings, J., de Matos, P., Rijnbeek, \n",
            "M., et al. (2012). MetaboLights—An open-access general-purpose \n",
            "repository for metabolomics studies and associated meta-data. \n",
            "Nucleic Acids Research, 41, D781–D786.\n",
            "Holten, D. (2006). Hierarchical edge bundles: Visualization of adja-\n",
            "cency relations in hierarchical data. IEEE Transactions on Visu-\n",
            "alization and Computer Graphics, 12, 741–748.\n",
            "Horai, H., Arita, M., Kanaya, S., Nihei, Y., Ikeda, T., Suwa, K., et al. \n",
            "(2010). MassBank: A public repository for sharing mass spectral \n",
            "data for life sciences. Journal of Mass Spectrometry, 45, 703–714.\n",
            "Hunter, J. D. (2007). Matplotlib: A 2D graphics environment. Comput-\n",
            "ing in Science & Engineering, 9, 90–95.\n",
            "Jones, M. (2013). Python for biologists. Scotts Valley: CreateSpace \n",
            "Independent Publishing Platform.\n",
            "Kluyver, T., Ragan-Kelley, B., Pérez, F., Granger, B., Bussonnier, M., \n",
            "Frederic, J., et al. (2016). Jupyter Notebooks—a publishing format \n",
            "for reproducible computational workflows in Loizides. In F. A. \n",
            "S. Birgi (Ed.), Positioning and power in academic publishing: \n",
            "Players, agents and agendas (pp. 87–90). Amsterdam: IOS Press.\n",
            "Knuth, D. E. (1984). Literate programming. The Computer Journal, \n",
            "27, 97–111.\n",
            "Kolb, D. (1984). Experiential learning: Experience as the source of \n",
            "learning and development. Englewood Cliffs, NJ: Prentice Hall.\n",
            "Lantz, B. (2013). Machine learning with R (1st ed.). Birmingham: \n",
            "Packt Publishing.\n",
            "Lee, A. H., Shannon, C. P., Amenyogbe, N., Bennike, T. B., Diray-\n",
            "Arce, J., Idoko, O. T., et al. (2019). Dynamic molecular changes \n",
            "during the first week of human life follow a robust developmental \n",
            "trajectory. Nature Communications, 10, 1092.\n",
            "McKinney, W. (2017). Python for data analysis (2nd ed.). Newton: \n",
            "O’Reilly Media Inc.\n",
            "Müller, A. C., & Guido, S. (2017). Introduction to machine learn-\n",
            "ing with Python: A guide for data scientists (1st ed.). California: \n",
            "O’Reilly Media Inc.\n",
            "Open Science MOOC. (2018). Make your code citable using GitHub \n",
            "and Zenodo: A how-to guide. Retrieved August 14, 2019, from \n",
            "https ://genr.eu/wp/cite/.\n",
            "Passey, D. (2017). Computer science (CS) in the compulsory educa-\n",
            "tion curriculum: Implications for future research. Education and \n",
            "Information Technologies, 22, 421–443.\n",
            "Pedregosa, A. F., Varoquaux, A. G., Gramfort, A. A., Michel, A. V., \n",
            "Thirion, A. B., Grisel, A. O., et al. (2011). Scikit-learn: Machine \n",
            "learning in Python. Journal of Machine Learning Research, 12, \n",
            "2825–2830.\n",
            "Pedrioli, P. G., Eng, J. K., Hubley, R., Vogelzang, M., Deutsch, E. \n",
            "W., Raught, B., et al. (2004). A common open representation of \n",
            "mass spectrometry data and its application to proteomics research. \n",
            "Nature Biotechnology, 22, 1459–1466.\n",
            "Peters, K., Bradbury, J., Bergmann, S., Capuccini, M., Cascante, M., de \n",
            "Atauri, P., et al. (2019). PhenoMeNal: processing and analysis of \n",
            "metabolomics data in the cloud. GigaScience, 8, giy149.\n",
            "Pinu, F. R., Beale, D. J., Paten, A. M., Kouremenos, K., Swarup, S., \n",
            "Schirra, H. J., et al. (2019). Systems biology and multi-omics inte-\n",
            "gration: Viewpoints from the metabolomics research community. \n",
            "Metabolites, 9, 76.\n",
            "Project Jupyter (2019). Jupyter. Retrieved March 19, 2019, from https \n",
            "://jupyt er.org/.\n",
            "Project Jupyter, Bussonnier, M., Forde, J., Freeman, J., Granger, \n",
            "B., Head, T., Holdgraf, C., Kelley, K., Nalvarte, G., Osheroff, \n",
            "A., Pacer, M., Panda, Y., Perez, F., Ragan-Kelley, B. and Will-\n",
            "ing, C. (2018) Binder 2.0—Reproducible, interactive, sharable \n",
            "environments for science at scale, SCIPY 2018, Proceedings of \n",
            "the 17th Python in Science Conference, pp. 113–120.\n",
            "Ramalho, L. (2015). Fluent python: Clear, concise, and effective pro-\n",
            "gramming. Sebastopol, CA: O’Reilly Media Inc.\n",
            "Reinke, S. N., Galindo-Prieto, B., Skotare, T., Broadhurst, D. I., \n",
            "Singhania, A., Horowitz, D., et al. (2018). OnPLS-based multi-\n",
            "block data integration: A multivariate approach to interrogat-\n",
            "ing biological interactions in asthma. Analytical Chemistry, 90, \n",
            "13400–13408.\n",
            "Rohart, F., Gautier, B., Singh, A., & Lê Cao, K.-A. (2017). mixOm-\n",
            "ics: An R package for ‘omics feature selection and multiple data \n",
            "integration. PLoS Computational Biology, 13, e1005752.\n",
            "Sicilia, M.-A., García-Barriocanal, E., & Sánchez-Alonso, S. (2017). \n",
            "Community curation in open dataset repositories: Insights from \n",
            "Zenodo. Procedia Computer Science, 106, 54–60.\n",
            "Smith, C. A., O’Maille, G., Want, E. J., Qin, C., Trauger, S. A., Bran-\n",
            "don, T. R., et al. (2005). METLIN: A metabolite mass spectral \n",
            "database. Therapeutic Drug Monitoring, 27, 747–751.\n",
            "Spicer, R. A., Salek, R., & Steinbeck, C. (2017). A decade after the \n",
            "metabolomics standards initiative it’s time for a revision. Scientific \n",
            "Data, 4, 170138.\n",
            "Sud, M., Fahy, E., Cotter, D., Azam, K., Vadivelu, I., Burant, C., et al. \n",
            "(2016). Metabolomics Workbench: An international repository for \n",
            "metabolomics data and metadata, metabolite standards, protocols, \n",
            "tutorials and training, and analysis tools. Nucleic Acids Research, \n",
            "44, D463–D470.\n",
            "Teschendorff, A. E. (2019). Avoiding common pitfalls in machine \n",
            "learning omic data science. Nature Materials, 18, 422–427.\n",
            "The Carpentries (2019). Lessons. Retrieved May 20, 2019, from https \n",
            "://softw are-carpe ntry.org/lesso ns/.\n",
            "van der Walt, S., Colbert, S. C., & Varoquaux, G. (2011). The NumPy \n",
            "array: A structure for efficient numerical computation. Computing \n",
            "in Science & Engineering, 13, 22–30.\n",
            "VanderPlas, J. (2016). Python data science handbook: Essential tools \n",
            "for working with data. Sebastopol, CA: O’Reilly Media Inc.\n",
            "Waskom, M., Botvinnik, O., O’Kane, D., Hobson, P., Ostblom, J., \n",
            "Lukauskas, S., Gemperline, D.C., Augspurger, T., Halchenko, \n",
            "Y., Cole, J.B., Warmenhoven, J., Ruiter, J.d., Pye, C., Hoyer, S., \n",
            "Vanderplas, J., Villalba, S., Kunter, G., Quintero, E., Bachant, P., \n",
            "Martin, M., Meyer, K., Miles, A., Ram, Y., Brunner, T., Yarkoni, \n",
            "T., Williams, M.L., Evans, C., Fitzgerald, C., Brian and Qalieh, A. \n",
            "(2018). mwaskom/seaborn: v0.9.0. Retrieved May 1, 2019, from \n",
            "https ://doi.org/10.5281/zenod o.13132 01.\n",
            "Wickham, H. (2014). Tidy data. Journal of Statistical Software, 59, \n",
            "1–23.\n",
            "Wilkinson, M. D., Dumontier, M., Aalbersberg, I. J., Appleton, G., \n",
            "Axton, M., Baak, A., et al. (2016). The FAIR guiding principles \n",
            "for scientific data management and stewardship. Scientific Data, \n",
            "3, 160018.\n",
            "Wishart, D. S., Feunang, Y. D., Marcu, A., Guo, A. C., Liang, K., \n",
            "Vazquez-Fresno, R., et  al. (2018). HMDB 4.0: The human \n",
            "metabolome database for 2018. Nucleic Acids Research, 46, \n",
            "D608–D617.\n",
            "Xia, J., Broadhurst, D. I., Wilson, M., & Wishart, D. S. (2013). Transla-\n",
            "tional biomarker discovery in clinical metabolomics: An introduc-\n",
            "tory tutorial. Metabolomics, 9, 280–299.\n",
            "Xia, J., & Wishart, D. S. (2011). Metabolomic data processing, analy-\n",
            "sis, and interpretation using MetaboAnalyst. Current Protocols in \n",
            "Bioinformatics, 34(1), 10–14.\n",
            "Publisher’s Note Springer Nature remains neutral with regard to \n",
            "jurisdictional claims in published maps and institutional affiliations.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pdf_summary = summarizer(text[:1024], min_length =100, max_length=250)\n",
        "pdf_summary"
      ],
      "metadata": {
        "id": "YgIkX3aafZeg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "43fec47f-aa43-4990-8c69-b9ba6c128cd9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[{'summary_text': ' The metabolomics community has made substantial efforts to align with FAIR data standards by promoting open data formats, online spectral libraries, and metabolite databases . A lack of transparency and reporting standards in the scientific community has led to increasing and widespread concerns relating to reproduction and integrity of results . As an omics science, metabolomics is highly vulnerable to irreproducibility. It generates vast amounts of data and relies heavily on data science for deriving biological meaning . The community is encouraging open data format, open data repositories, online . databases, and open data analysis platforms .'}]"
            ]
          },
          "metadata": {},
          "execution_count": 19
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Summarise Webpage\n",
        "\n",
        "\n",
        "https://www.geeksforgeeks.org/implementing-web-scraping-python-beautiful-soup/\n",
        "\n"
      ],
      "metadata": {
        "id": "E7c-hpWGMjUc"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "NqvjK_TmMqkT"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}